{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n",
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import os, sys\n",
    "from IPython.display import Image\n",
    "import pandas as pd\n",
    "from __future__ import division\n",
    "import numpy as np\n",
    "import rpy2\n",
    "from rpy2 import robjects as ro\n",
    "import pandas.rpy.common as com\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import operator\n",
    "import scipy as sp\n",
    "import traceback\n",
    "from sklearn import preprocessing\n",
    "from IPython.parallel import Client\n",
    "from subprocess import Popen, PIPE\n",
    "import shutil\n",
    "from IPython.display import FileLink, FileLinks, Image\n",
    "import psutil\n",
    "import multiprocessing\n",
    "from hdfstorehelper import HDFStoreHelper\n",
    "import warnings\n",
    "import pandas\n",
    "import dill\n",
    "import statsmodels as sm\n",
    "import statsmodels.formula.api as smf\n",
    "from scipy.stats.stats import pearsonr\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "warnings.simplefilter(\"ignore\", pandas.io.pytables.PerformanceWarning)\n",
    "%matplotlib inline\n",
    "\n",
    "%load_ext rpy2.ipython\n",
    "pd.set_option('display.width', 80)\n",
    "pd.set_option('max.columns', 30)\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "sns.set_context(\"talk\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def convert_to_snpassoc(col):\n",
    "    if \"-\" in col.name:\n",
    "        freqs = af[col.name]\n",
    "        trans = {11: \"%s/%s\" % (freqs[\"A\"], freqs[\"A\"]),\n",
    "                12: \"%s/%s\" % (freqs[\"A\"], freqs[\"a\"]),\n",
    "                22: \"%s/%s\" % (freqs[\"a\"], freqs[\"a\"]),\n",
    "                \"NA\":\"NA\"}\n",
    "        return col.apply(lambda x: trans[x])\n",
    "    return col\n",
    "\n",
    "def get_phenotype(row):\n",
    "    return np.max(pheno[(pheno.Longitude==row.long) & (pheno.Latitude==row.lat)])\n",
    "\n",
    "def center_and_standardize_value(val, u, var):\n",
    "    if val == -1:\n",
    "        return 0.0\n",
    "    return (val-u)/np.sqrt(var)\n",
    "\n",
    "def center_and_standardize(snp):\n",
    "    maf = af.ix[\"q\",snp.name]\n",
    "    u = np.mean([x for x in snp if x != -1])\n",
    "    var = np.sqrt(maf*(1-maf))\n",
    "    return snp.apply(center_and_standardize_value, args=(u, var))\n",
    "\n",
    "def add_county_id(row):\n",
    "    key = \"%s_%s\" % (row.county,row.state)\n",
    "    if key in county_id:\n",
    "        return county_id[key]\n",
    "    return np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "r = ro.r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'/gdc_home4/cfried/ipython'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hdf = HDFStoreHelper(\"cfried.hd5\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_name = dill.load(open(\"cfried_trait_name.dill\"))\n",
    "county_id = dill.load(open(\"county_id.dill\"))\n",
    "pca_cov = hdf.get('pca_cov')\n",
    "loc_hierf = hdf.get(\"loc_hierf\")\n",
    "pheno = hdf.get('pheno')\n",
    "af = hdf.get(\"af\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait = loc_hierf.apply(get_phenotype, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "trait_loc_hierf = trait.join(loc_hierf, how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_complete = trait_loc_hierf.drop(trait_loc_hierf[np.isnan(trait_loc_hierf[trait_name])].index)\n",
    "trait_complete['countyid'] = trait_complete.apply(add_county_id, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>melezitose</th>\n",
       "      <th>county</th>\n",
       "      <th>state</th>\n",
       "      <th>lat</th>\n",
       "      <th>long</th>\n",
       "      <th>county_state</th>\n",
       "      <th>countyid</th>\n",
       "      <th>0-10037-01-257</th>\n",
       "      <th>0-10040-02-394</th>\n",
       "      <th>0-10044-01-392</th>\n",
       "      <th>0-10048-01-60</th>\n",
       "      <th>0-10051-02-166</th>\n",
       "      <th>0-10054-01-402</th>\n",
       "      <th>...</th>\n",
       "      <th>UMN-CL299Contig1-01-46</th>\n",
       "      <th>UMN-CL306Contig1-04-261</th>\n",
       "      <th>UMN-CL307Contig1-04-143</th>\n",
       "      <th>UMN-CL319Contig1-03-131</th>\n",
       "      <th>UMN-CL326Contig1-05-421</th>\n",
       "      <th>UMN-CL339Contig1-05-39</th>\n",
       "      <th>UMN-CL34Contig1-03-89</th>\n",
       "      <th>UMN-CL353Contig1-04-64</th>\n",
       "      <th>UMN-CL362Contig1-07-133</th>\n",
       "      <th>UMN-CL363Contig1-01-233</th>\n",
       "      <th>UMN-CL379Contig1-12-117</th>\n",
       "      <th>UMN-CL424Contig1-03-94</th>\n",
       "      <th>UMN-CL54Contig1-07-88</th>\n",
       "      <th>UMN-CL91Contig1-02-246</th>\n",
       "      <th>UMN-CL97Contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-76.87746</td>\n",
       "      <td> 37.66986</td>\n",
       "      <td> 3.182480</td>\n",
       "      <td> KING &amp; QUEEN</td>\n",
       "      <td> VA</td>\n",
       "      <td> 37.66986</td>\n",
       "      <td>-76.87746</td>\n",
       "      <td> KING &amp; QUEEN_VA</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 11</td>\n",
       "      <td> 22</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td>...</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> NA</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-76.87746</td>\n",
       "      <td> 37.66986</td>\n",
       "      <td> 3.182480</td>\n",
       "      <td> KING &amp; QUEEN</td>\n",
       "      <td> VA</td>\n",
       "      <td> 37.66986</td>\n",
       "      <td>-76.87746</td>\n",
       "      <td> KING &amp; QUEEN_VA</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> NA</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td>...</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-76.97319</td>\n",
       "      <td> 37.51160</td>\n",
       "      <td> 2.946809</td>\n",
       "      <td>     NEW KENT</td>\n",
       "      <td> VA</td>\n",
       "      <td> 37.51160</td>\n",
       "      <td>-76.97319</td>\n",
       "      <td>     NEW KENT_VA</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 22</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td>...</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> NA</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>-77.42219</td>\n",
       "      <td> 36.39032</td>\n",
       "      <td> 3.248399</td>\n",
       "      <td>  NORTHAMPTON</td>\n",
       "      <td> NC</td>\n",
       "      <td> 36.39032</td>\n",
       "      <td>-77.42219</td>\n",
       "      <td>  NORTHAMPTON_NC</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 12</td>\n",
       "      <td> 12</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 12</td>\n",
       "      <td>...</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-78.70453</td>\n",
       "      <td> 34.33010</td>\n",
       "      <td> 3.359321</td>\n",
       "      <td>     COLUMBUS</td>\n",
       "      <td> NC</td>\n",
       "      <td> 34.33010</td>\n",
       "      <td>-78.70453</td>\n",
       "      <td>     COLUMBUS_NC</td>\n",
       "      <td> 11</td>\n",
       "      <td> NA</td>\n",
       "      <td> 22</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td>...</td>\n",
       "      <td> 11</td>\n",
       "      <td> NA</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> NA</td>\n",
       "      <td> 11</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 11</td>\n",
       "      <td> 12</td>\n",
       "      <td> 22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3091 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Longitude  Latitude  melezitose        county state       lat      long  \\\n",
       "3  -76.87746  37.66986    3.182480  KING & QUEEN    VA  37.66986 -76.87746   \n",
       "4  -76.87746  37.66986    3.182480  KING & QUEEN    VA  37.66986 -76.87746   \n",
       "5  -76.97319  37.51160    2.946809      NEW KENT    VA  37.51160 -76.97319   \n",
       "7  -77.42219  36.39032    3.248399   NORTHAMPTON    NC  36.39032 -77.42219   \n",
       "8  -78.70453  34.33010    3.359321      COLUMBUS    NC  34.33010 -78.70453   \n",
       "\n",
       "      county_state  countyid 0-10037-01-257 0-10040-02-394 0-10044-01-392  \\\n",
       "3  KING & QUEEN_VA       NaN             11             22             11   \n",
       "4  KING & QUEEN_VA       NaN             11             12             11   \n",
       "5      NEW KENT_VA       NaN             12             11             12   \n",
       "7   NORTHAMPTON_NC       NaN             12             12             12   \n",
       "8      COLUMBUS_NC        11             NA             22             11   \n",
       "\n",
       "  0-10048-01-60 0-10051-02-166 0-10054-01-402      ...        \\\n",
       "3            11             12             11      ...         \n",
       "4            NA             11             12      ...         \n",
       "5            22             11             12      ...         \n",
       "7            11             12             12      ...         \n",
       "8            11             11             11      ...         \n",
       "\n",
       "  UMN-CL299Contig1-01-46 UMN-CL306Contig1-04-261 UMN-CL307Contig1-04-143  \\\n",
       "3                     11                      11                      11   \n",
       "4                     11                      11                      12   \n",
       "5                     11                      11                      12   \n",
       "7                     11                      11                      11   \n",
       "8                     11                      NA                      12   \n",
       "\n",
       "  UMN-CL319Contig1-03-131 UMN-CL326Contig1-05-421 UMN-CL339Contig1-05-39  \\\n",
       "3                      11                      11                     11   \n",
       "4                      11                      11                     11   \n",
       "5                      11                      12                     11   \n",
       "7                      12                      12                     11   \n",
       "8                      11                      11                     11   \n",
       "\n",
       "  UMN-CL34Contig1-03-89 UMN-CL353Contig1-04-64 UMN-CL362Contig1-07-133  \\\n",
       "3                    12                     11                      NA   \n",
       "4                    12                     11                      11   \n",
       "5                    12                     11                      NA   \n",
       "7                    12                     11                      12   \n",
       "8                    11                     11                      NA   \n",
       "\n",
       "  UMN-CL363Contig1-01-233 UMN-CL379Contig1-12-117 UMN-CL424Contig1-03-94  \\\n",
       "3                      12                      11                     12   \n",
       "4                      12                      11                     12   \n",
       "5                      12                      11                     12   \n",
       "7                      12                      11                     11   \n",
       "8                      11                      11                     12   \n",
       "\n",
       "  UMN-CL54Contig1-07-88 UMN-CL91Contig1-02-246 UMN-CL97Contig  \n",
       "3                    12                     11             11  \n",
       "4                    11                     12             11  \n",
       "5                    11                     12             12  \n",
       "7                    12                     11             22  \n",
       "8                    11                     12             22  \n",
       "\n",
       "[5 rows x 3091 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_complete.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "trait_complete = trait_complete[trait_complete.countyid > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(330, 3091)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_complete.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_snpassoc = trait_complete.apply(convert_to_snpassoc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_snpassoc_pca = trait_snpassoc.join(pca_cov, how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>melezitose</th>\n",
       "      <th>county</th>\n",
       "      <th>state</th>\n",
       "      <th>lat</th>\n",
       "      <th>long</th>\n",
       "      <th>county_state</th>\n",
       "      <th>countyid</th>\n",
       "      <th>0-10037-01-257</th>\n",
       "      <th>0-10040-02-394</th>\n",
       "      <th>0-10044-01-392</th>\n",
       "      <th>0-10048-01-60</th>\n",
       "      <th>0-10051-02-166</th>\n",
       "      <th>0-10054-01-402</th>\n",
       "      <th>...</th>\n",
       "      <th>UMN-CL97Contig</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "      <th>PC6</th>\n",
       "      <th>PC7</th>\n",
       "      <th>PC8</th>\n",
       "      <th>PC9</th>\n",
       "      <th>PC10</th>\n",
       "      <th>PC11</th>\n",
       "      <th>PC12</th>\n",
       "      <th>PC13</th>\n",
       "      <th>PC14</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8 </th>\n",
       "      <td>-78.70453</td>\n",
       "      <td> 34.33010</td>\n",
       "      <td> 3.359321</td>\n",
       "      <td>   COLUMBUS</td>\n",
       "      <td> NC</td>\n",
       "      <td> 34.33010</td>\n",
       "      <td>-78.70453</td>\n",
       "      <td>   COLUMBUS_NC</td>\n",
       "      <td> 11</td>\n",
       "      <td>  NA</td>\n",
       "      <td> A/A</td>\n",
       "      <td> C/C</td>\n",
       "      <td> G/G</td>\n",
       "      <td> G/G</td>\n",
       "      <td> G/G</td>\n",
       "      <td>...</td>\n",
       "      <td> A/A</td>\n",
       "      <td>-3.639308</td>\n",
       "      <td> 1.929420</td>\n",
       "      <td> 1.193841</td>\n",
       "      <td>-3.447265</td>\n",
       "      <td> 0.971588</td>\n",
       "      <td> 1.904022</td>\n",
       "      <td>-2.019537</td>\n",
       "      <td> 4.332102</td>\n",
       "      <td>-2.889919</td>\n",
       "      <td> 1.682956</td>\n",
       "      <td>-6.441526</td>\n",
       "      <td> 4.149832</td>\n",
       "      <td>-0.223308</td>\n",
       "      <td>-0.780265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>-79.30539</td>\n",
       "      <td> 33.36318</td>\n",
       "      <td> 2.919807</td>\n",
       "      <td> GEORGETOWN</td>\n",
       "      <td> SC</td>\n",
       "      <td> 33.36318</td>\n",
       "      <td>-79.30539</td>\n",
       "      <td> GEORGETOWN_SC</td>\n",
       "      <td> 14</td>\n",
       "      <td>  NA</td>\n",
       "      <td> C/A</td>\n",
       "      <td>  NA</td>\n",
       "      <td>  NA</td>\n",
       "      <td> G/G</td>\n",
       "      <td> G/A</td>\n",
       "      <td>...</td>\n",
       "      <td>  NA</td>\n",
       "      <td> 2.329722</td>\n",
       "      <td> 1.624366</td>\n",
       "      <td> 0.335046</td>\n",
       "      <td>-0.417052</td>\n",
       "      <td>-1.013743</td>\n",
       "      <td>-0.138149</td>\n",
       "      <td> 1.068138</td>\n",
       "      <td> 0.402887</td>\n",
       "      <td>-0.010373</td>\n",
       "      <td> 0.559439</td>\n",
       "      <td> 2.445590</td>\n",
       "      <td> 0.335334</td>\n",
       "      <td>-0.047097</td>\n",
       "      <td>-1.344021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>-77.05205</td>\n",
       "      <td> 35.55349</td>\n",
       "      <td> 3.048517</td>\n",
       "      <td>   BEAUFORT</td>\n",
       "      <td> NC</td>\n",
       "      <td> 35.55349</td>\n",
       "      <td>-77.05205</td>\n",
       "      <td>   BEAUFORT_NC</td>\n",
       "      <td>  2</td>\n",
       "      <td> A/A</td>\n",
       "      <td>  NA</td>\n",
       "      <td> C/G</td>\n",
       "      <td> G/A</td>\n",
       "      <td> G/G</td>\n",
       "      <td> A/A</td>\n",
       "      <td>...</td>\n",
       "      <td> G/G</td>\n",
       "      <td>-2.708652</td>\n",
       "      <td> 3.668354</td>\n",
       "      <td> 4.708462</td>\n",
       "      <td> 0.815316</td>\n",
       "      <td> 5.354841</td>\n",
       "      <td> 5.688271</td>\n",
       "      <td> 6.544227</td>\n",
       "      <td> 1.869980</td>\n",
       "      <td>-0.408042</td>\n",
       "      <td> 2.274323</td>\n",
       "      <td> 0.627999</td>\n",
       "      <td>-6.249547</td>\n",
       "      <td>-8.086907</td>\n",
       "      <td> 8.983447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>-77.06917</td>\n",
       "      <td> 35.10917</td>\n",
       "      <td> 2.527383</td>\n",
       "      <td>     CRAVEN</td>\n",
       "      <td> NC</td>\n",
       "      <td> 35.10917</td>\n",
       "      <td>-77.06917</td>\n",
       "      <td>     CRAVEN_NC</td>\n",
       "      <td> 12</td>\n",
       "      <td> A/G</td>\n",
       "      <td> C/A</td>\n",
       "      <td> C/G</td>\n",
       "      <td> G/A</td>\n",
       "      <td> G/G</td>\n",
       "      <td> G/A</td>\n",
       "      <td>...</td>\n",
       "      <td> G/G</td>\n",
       "      <td>-3.152409</td>\n",
       "      <td> 2.664518</td>\n",
       "      <td> 2.302811</td>\n",
       "      <td>-0.078610</td>\n",
       "      <td> 0.237071</td>\n",
       "      <td>-3.257392</td>\n",
       "      <td>-2.636748</td>\n",
       "      <td> 3.809404</td>\n",
       "      <td>-0.443443</td>\n",
       "      <td> 7.257342</td>\n",
       "      <td>-0.466697</td>\n",
       "      <td> 1.824831</td>\n",
       "      <td>-0.198020</td>\n",
       "      <td>-1.268526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>-77.05205</td>\n",
       "      <td> 35.55349</td>\n",
       "      <td> 3.048517</td>\n",
       "      <td>   BEAUFORT</td>\n",
       "      <td> NC</td>\n",
       "      <td> 35.55349</td>\n",
       "      <td>-77.05205</td>\n",
       "      <td>   BEAUFORT_NC</td>\n",
       "      <td>  2</td>\n",
       "      <td> A/A</td>\n",
       "      <td> C/C</td>\n",
       "      <td> G/G</td>\n",
       "      <td> G/G</td>\n",
       "      <td> G/A</td>\n",
       "      <td> G/A</td>\n",
       "      <td>...</td>\n",
       "      <td> G/G</td>\n",
       "      <td>-3.806482</td>\n",
       "      <td> 2.723616</td>\n",
       "      <td>-5.248239</td>\n",
       "      <td>-2.665990</td>\n",
       "      <td> 4.428649</td>\n",
       "      <td>-2.327034</td>\n",
       "      <td>-2.104905</td>\n",
       "      <td> 5.450938</td>\n",
       "      <td>-3.403810</td>\n",
       "      <td>-0.727319</td>\n",
       "      <td> 4.328464</td>\n",
       "      <td>-1.099518</td>\n",
       "      <td> 0.683061</td>\n",
       "      <td> 0.534313</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3105 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Longitude  Latitude  melezitose      county state       lat      long  \\\n",
       "8   -78.70453  34.33010    3.359321    COLUMBUS    NC  34.33010 -78.70453   \n",
       "12  -79.30539  33.36318    2.919807  GEORGETOWN    SC  33.36318 -79.30539   \n",
       "13  -77.05205  35.55349    3.048517    BEAUFORT    NC  35.55349 -77.05205   \n",
       "15  -77.06917  35.10917    2.527383      CRAVEN    NC  35.10917 -77.06917   \n",
       "18  -77.05205  35.55349    3.048517    BEAUFORT    NC  35.55349 -77.05205   \n",
       "\n",
       "     county_state  countyid 0-10037-01-257 0-10040-02-394 0-10044-01-392  \\\n",
       "8     COLUMBUS_NC        11             NA            A/A            C/C   \n",
       "12  GEORGETOWN_SC        14             NA            C/A             NA   \n",
       "13    BEAUFORT_NC         2            A/A             NA            C/G   \n",
       "15      CRAVEN_NC        12            A/G            C/A            C/G   \n",
       "18    BEAUFORT_NC         2            A/A            C/C            G/G   \n",
       "\n",
       "   0-10048-01-60 0-10051-02-166 0-10054-01-402    ...    UMN-CL97Contig  \\\n",
       "8            G/G            G/G            G/G    ...               A/A   \n",
       "12            NA            G/G            G/A    ...                NA   \n",
       "13           G/A            G/G            A/A    ...               G/G   \n",
       "15           G/A            G/G            G/A    ...               G/G   \n",
       "18           G/G            G/A            G/A    ...               G/G   \n",
       "\n",
       "         PC1       PC2       PC3       PC4       PC5       PC6       PC7  \\\n",
       "8  -3.639308  1.929420  1.193841 -3.447265  0.971588  1.904022 -2.019537   \n",
       "12  2.329722  1.624366  0.335046 -0.417052 -1.013743 -0.138149  1.068138   \n",
       "13 -2.708652  3.668354  4.708462  0.815316  5.354841  5.688271  6.544227   \n",
       "15 -3.152409  2.664518  2.302811 -0.078610  0.237071 -3.257392 -2.636748   \n",
       "18 -3.806482  2.723616 -5.248239 -2.665990  4.428649 -2.327034 -2.104905   \n",
       "\n",
       "         PC8       PC9      PC10      PC11      PC12      PC13      PC14  \n",
       "8   4.332102 -2.889919  1.682956 -6.441526  4.149832 -0.223308 -0.780265  \n",
       "12  0.402887 -0.010373  0.559439  2.445590  0.335334 -0.047097 -1.344021  \n",
       "13  1.869980 -0.408042  2.274323  0.627999 -6.249547 -8.086907  8.983447  \n",
       "15  3.809404 -0.443443  7.257342 -0.466697  1.824831 -0.198020 -1.268526  \n",
       "18  5.450938 -3.403810 -0.727319  4.328464 -1.099518  0.683061  0.534313  \n",
       "\n",
       "[5 rows x 3105 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_snpassoc_pca.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_snpassoc_pca = trait_snpassoc_pca.drop(['county_state',\n",
    "                         'Longitude',\n",
    "                         'Latitude',\n",
    "                         'county',\n",
    "                         'state',\n",
    "                         'lat',\n",
    "                         'long',\n",
    "                         'countyid'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>melezitose</th>\n",
       "      <th>0-10037-01-257</th>\n",
       "      <th>0-10040-02-394</th>\n",
       "      <th>0-10044-01-392</th>\n",
       "      <th>0-10048-01-60</th>\n",
       "      <th>0-10051-02-166</th>\n",
       "      <th>0-10054-01-402</th>\n",
       "      <th>0-10067-03-111</th>\n",
       "      <th>0-10079-02-168</th>\n",
       "      <th>0-10112-01-169</th>\n",
       "      <th>0-10113-01-119</th>\n",
       "      <th>0-10116-01-165</th>\n",
       "      <th>0-10151-01-86</th>\n",
       "      <th>0-10162-01-255</th>\n",
       "      <th>0-10207-01-280</th>\n",
       "      <th>...</th>\n",
       "      <th>UMN-CL97Contig</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "      <th>PC6</th>\n",
       "      <th>PC7</th>\n",
       "      <th>PC8</th>\n",
       "      <th>PC9</th>\n",
       "      <th>PC10</th>\n",
       "      <th>PC11</th>\n",
       "      <th>PC12</th>\n",
       "      <th>PC13</th>\n",
       "      <th>PC14</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8 </th>\n",
       "      <td> 3.359321</td>\n",
       "      <td>  NA</td>\n",
       "      <td> A/A</td>\n",
       "      <td> C/C</td>\n",
       "      <td> G/G</td>\n",
       "      <td> G/G</td>\n",
       "      <td> G/G</td>\n",
       "      <td> A/A</td>\n",
       "      <td> G/G</td>\n",
       "      <td> A/A</td>\n",
       "      <td> A/G</td>\n",
       "      <td>  NA</td>\n",
       "      <td> A/A</td>\n",
       "      <td> A/G</td>\n",
       "      <td> A/C</td>\n",
       "      <td>...</td>\n",
       "      <td> A/A</td>\n",
       "      <td>-3.639308</td>\n",
       "      <td> 1.929420</td>\n",
       "      <td> 1.193841</td>\n",
       "      <td>-3.447265</td>\n",
       "      <td> 0.971588</td>\n",
       "      <td> 1.904022</td>\n",
       "      <td>-2.019537</td>\n",
       "      <td> 4.332102</td>\n",
       "      <td>-2.889919</td>\n",
       "      <td> 1.682956</td>\n",
       "      <td>-6.441526</td>\n",
       "      <td> 4.149832</td>\n",
       "      <td>-0.223308</td>\n",
       "      <td>-0.780265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td> 2.919807</td>\n",
       "      <td>  NA</td>\n",
       "      <td> C/A</td>\n",
       "      <td>  NA</td>\n",
       "      <td>  NA</td>\n",
       "      <td> G/G</td>\n",
       "      <td> G/A</td>\n",
       "      <td> A/A</td>\n",
       "      <td> G/G</td>\n",
       "      <td> A/C</td>\n",
       "      <td> A/G</td>\n",
       "      <td> A/C</td>\n",
       "      <td> A/A</td>\n",
       "      <td> A/A</td>\n",
       "      <td> A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>  NA</td>\n",
       "      <td> 2.329722</td>\n",
       "      <td> 1.624366</td>\n",
       "      <td> 0.335046</td>\n",
       "      <td>-0.417052</td>\n",
       "      <td>-1.013743</td>\n",
       "      <td>-0.138149</td>\n",
       "      <td> 1.068138</td>\n",
       "      <td> 0.402887</td>\n",
       "      <td>-0.010373</td>\n",
       "      <td> 0.559439</td>\n",
       "      <td> 2.445590</td>\n",
       "      <td> 0.335334</td>\n",
       "      <td>-0.047097</td>\n",
       "      <td>-1.344021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td> 3.048517</td>\n",
       "      <td> A/A</td>\n",
       "      <td>  NA</td>\n",
       "      <td> C/G</td>\n",
       "      <td> G/A</td>\n",
       "      <td> G/G</td>\n",
       "      <td> A/A</td>\n",
       "      <td> A/A</td>\n",
       "      <td> G/G</td>\n",
       "      <td> A/A</td>\n",
       "      <td> A/G</td>\n",
       "      <td> A/A</td>\n",
       "      <td> A/A</td>\n",
       "      <td> A/A</td>\n",
       "      <td> C/C</td>\n",
       "      <td>...</td>\n",
       "      <td> G/G</td>\n",
       "      <td>-2.708652</td>\n",
       "      <td> 3.668354</td>\n",
       "      <td> 4.708462</td>\n",
       "      <td> 0.815316</td>\n",
       "      <td> 5.354841</td>\n",
       "      <td> 5.688271</td>\n",
       "      <td> 6.544227</td>\n",
       "      <td> 1.869980</td>\n",
       "      <td>-0.408042</td>\n",
       "      <td> 2.274323</td>\n",
       "      <td> 0.627999</td>\n",
       "      <td>-6.249547</td>\n",
       "      <td>-8.086907</td>\n",
       "      <td> 8.983447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td> 2.527383</td>\n",
       "      <td> A/G</td>\n",
       "      <td> C/A</td>\n",
       "      <td> C/G</td>\n",
       "      <td> G/A</td>\n",
       "      <td> G/G</td>\n",
       "      <td> G/A</td>\n",
       "      <td> A/A</td>\n",
       "      <td> G/G</td>\n",
       "      <td> A/A</td>\n",
       "      <td> G/G</td>\n",
       "      <td> A/C</td>\n",
       "      <td> A/T</td>\n",
       "      <td> A/A</td>\n",
       "      <td> C/C</td>\n",
       "      <td>...</td>\n",
       "      <td> G/G</td>\n",
       "      <td>-3.152409</td>\n",
       "      <td> 2.664518</td>\n",
       "      <td> 2.302811</td>\n",
       "      <td>-0.078610</td>\n",
       "      <td> 0.237071</td>\n",
       "      <td>-3.257392</td>\n",
       "      <td>-2.636748</td>\n",
       "      <td> 3.809404</td>\n",
       "      <td>-0.443443</td>\n",
       "      <td> 7.257342</td>\n",
       "      <td>-0.466697</td>\n",
       "      <td> 1.824831</td>\n",
       "      <td>-0.198020</td>\n",
       "      <td>-1.268526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td> 3.048517</td>\n",
       "      <td> A/A</td>\n",
       "      <td> C/C</td>\n",
       "      <td> G/G</td>\n",
       "      <td> G/G</td>\n",
       "      <td> G/A</td>\n",
       "      <td> G/A</td>\n",
       "      <td> A/A</td>\n",
       "      <td> G/G</td>\n",
       "      <td> A/A</td>\n",
       "      <td> A/G</td>\n",
       "      <td> A/A</td>\n",
       "      <td> A/A</td>\n",
       "      <td> A/A</td>\n",
       "      <td> A/A</td>\n",
       "      <td>...</td>\n",
       "      <td> G/G</td>\n",
       "      <td>-3.806482</td>\n",
       "      <td> 2.723616</td>\n",
       "      <td>-5.248239</td>\n",
       "      <td>-2.665990</td>\n",
       "      <td> 4.428649</td>\n",
       "      <td>-2.327034</td>\n",
       "      <td>-2.104905</td>\n",
       "      <td> 5.450938</td>\n",
       "      <td>-3.403810</td>\n",
       "      <td>-0.727319</td>\n",
       "      <td> 4.328464</td>\n",
       "      <td>-1.099518</td>\n",
       "      <td> 0.683061</td>\n",
       "      <td> 0.534313</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3097 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    melezitose 0-10037-01-257 0-10040-02-394 0-10044-01-392 0-10048-01-60  \\\n",
       "8     3.359321             NA            A/A            C/C           G/G   \n",
       "12    2.919807             NA            C/A             NA            NA   \n",
       "13    3.048517            A/A             NA            C/G           G/A   \n",
       "15    2.527383            A/G            C/A            C/G           G/A   \n",
       "18    3.048517            A/A            C/C            G/G           G/G   \n",
       "\n",
       "   0-10051-02-166 0-10054-01-402 0-10067-03-111 0-10079-02-168 0-10112-01-169  \\\n",
       "8             G/G            G/G            A/A            G/G            A/A   \n",
       "12            G/G            G/A            A/A            G/G            A/C   \n",
       "13            G/G            A/A            A/A            G/G            A/A   \n",
       "15            G/G            G/A            A/A            G/G            A/A   \n",
       "18            G/A            G/A            A/A            G/G            A/A   \n",
       "\n",
       "   0-10113-01-119 0-10116-01-165 0-10151-01-86 0-10162-01-255 0-10207-01-280  \\\n",
       "8             A/G             NA           A/A            A/G            A/C   \n",
       "12            A/G            A/C           A/A            A/A            A/A   \n",
       "13            A/G            A/A           A/A            A/A            C/C   \n",
       "15            G/G            A/C           A/T            A/A            C/C   \n",
       "18            A/G            A/A           A/A            A/A            A/A   \n",
       "\n",
       "      ...    UMN-CL97Contig       PC1       PC2       PC3       PC4       PC5  \\\n",
       "8     ...               A/A -3.639308  1.929420  1.193841 -3.447265  0.971588   \n",
       "12    ...                NA  2.329722  1.624366  0.335046 -0.417052 -1.013743   \n",
       "13    ...               G/G -2.708652  3.668354  4.708462  0.815316  5.354841   \n",
       "15    ...               G/G -3.152409  2.664518  2.302811 -0.078610  0.237071   \n",
       "18    ...               G/G -3.806482  2.723616 -5.248239 -2.665990  4.428649   \n",
       "\n",
       "         PC6       PC7       PC8       PC9      PC10      PC11      PC12  \\\n",
       "8   1.904022 -2.019537  4.332102 -2.889919  1.682956 -6.441526  4.149832   \n",
       "12 -0.138149  1.068138  0.402887 -0.010373  0.559439  2.445590  0.335334   \n",
       "13  5.688271  6.544227  1.869980 -0.408042  2.274323  0.627999 -6.249547   \n",
       "15 -3.257392 -2.636748  3.809404 -0.443443  7.257342 -0.466697  1.824831   \n",
       "18 -2.327034 -2.104905  5.450938 -3.403810 -0.727319  4.328464 -1.099518   \n",
       "\n",
       "        PC13      PC14  \n",
       "8  -0.223308 -0.780265  \n",
       "12 -0.047097 -1.344021  \n",
       "13 -8.086907  8.983447  \n",
       "15 -0.198020 -1.268526  \n",
       "18  0.683061  0.534313  \n",
       "\n",
       "[5 rows x 3097 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_snpassoc_pca.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "trait_snpassoc_pca.to_csv(\"cfried_%s_snpassoc.txt\" % trait_name,\n",
    "                             header=True,\n",
    "                             index=True,\n",
    "                             sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def write_snpassoc_file(df, input_file, num_pca_axes):\n",
    "    pheno = df.columns[0:1]\n",
    "    out_files = []\n",
    "    for p in pheno:\n",
    "        with open(\"snpassoc_%s_%s.R\" % (os.path.basename(input_file), p.lower()), \"w\") as o:\n",
    "            print \"writing %s\" % o.name\n",
    "            out_files.append(o.name)\n",
    "            text = '''\n",
    "library(SNPassoc)\n",
    "\n",
    "d = read.table('%s', sep=\"\\\\t\", row.names=1, header=T)\n",
    "\n",
    "#subtract b/c those are the PCA axes\n",
    "snp_cols = 2:(ncol(d)-%d)\n",
    "snp_data = setupSNP(d, colSNPs=snp_cols, sep=\"/\")\n",
    "pca_cols = (ncol(d)-%d):ncol(d)\n",
    "pca_data = d[,pca_cols]\n",
    "\n",
    "wg = WGassociation(%s~1+pca_data$PC1+pca_data$PC2+pca_data$PC3+pca_data$PC4+\n",
    "pca_data$PC5+pca_data$PC6+pca_data$PC7+pca_data$PC8+pca_data$PC9+pca_data$PC10+\n",
    "+pca_data$PC11+pca_data$PC12+pca_data$PC13+pca_data$PC14, \n",
    "data=snp_data, \n",
    "model=\"co\", \n",
    "genotypingRate=5)\n",
    "\n",
    "saveRDS(wg, \"wg_%s_co.rds\")\n",
    "stats = WGstats(wg)\n",
    "saveRDS(stats, \"wgstats_%s.rds\")\n",
    "''' % (input_file, \n",
    "       num_pca_axes,\n",
    "       num_pca_axes-1,\n",
    "       p, \n",
    "       p.lower(), \n",
    "       p.lower())\n",
    "        \n",
    "            o.write(text)\n",
    "    return out_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "writing snpassoc_cfried_melezitose_snpassoc.txt_melezitose.R\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[u'snpassoc_cfried_melezitose_snpassoc.txt_melezitose.R']"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "write_snpassoc_file(trait_snpassoc_pca, \"cfried_melezitose_snpassoc.txt\", 14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Run in R\n",
    "\n",
    "```R\n",
    "source('snpassoc_cfried_melezitose_snpassoc.txt_melezitose.R')\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "wg_trait_co.rds = readRDS('wg_melezitose_co.rds')\n",
    "wgstats_trait.rds = readRDS('wgstats_melezitose.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "wgstats_trait = r['wgstats_trait.rds']\n",
    "wgstats_trait_labels = r('labels(wg_trait_co.rds)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "converting melezitose\n"
     ]
    }
   ],
   "source": [
    "wgstats = {trait_name:[wgstats_trait, wgstats_trait_labels.rx2(1)]}\n",
    "for key, datalist in wgstats.items():\n",
    "    print \"converting %s\" % key\n",
    "    wgstats[key] = [com.convert_robj(x) for x in datalist]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_alleles(data):\n",
    "    a = set()\n",
    "    for x in data.index:\n",
    "        for elem in x.split(\"/\"):\n",
    "            a.add(elem)\n",
    "    return list(a)  \n",
    "\n",
    "def get_allele_freqs_wg(data, AA, Aa, aa):\n",
    "    total = np.sum(data['n'])*2\n",
    "    A = data.ix[AA, \"n\"]*2 + data.ix[Aa, \"n\"]\n",
    "    a = data.ix[aa, \"n\"]*2 + data.ix[Aa, \"n\"]\n",
    "    return A/total, a/total\n",
    "\n",
    "def get_genotypes(data, alleles):\n",
    "    homos = [\"%s/%s\" % (x,x) for x in alleles]\n",
    "    Aa = \"%s/%s\" % (alleles[0], alleles[1])\n",
    "    if Aa not in data.index:\n",
    "        Aa = Aa[::-1] #reverse it\n",
    "    AA, aa = homos\n",
    "    if data.ix[AA, \"n\"] < data.ix[aa, \"n\"]:\n",
    "        AA, aa = homos[::-1] #reverse it so that major is first\n",
    "    return AA, Aa, aa\n",
    "\n",
    "def get_genotypic_values(data, alleles):\n",
    "    AA, Aa, aa = get_genotypes(data, alleles)\n",
    "    G_AA = float(data.ix[AA, 'me'])\n",
    "    G_aa = float(data.ix[aa, 'me'])\n",
    "    additive = (G_AA-G_aa)/2\n",
    "    G_Aa = float(data.ix[Aa, 'me'])\n",
    "    dominance = G_Aa - ((G_AA+G_aa)/2)\n",
    "    return additive, dominance, AA, Aa, aa\n",
    "    \n",
    "def get_alpha(data):\n",
    "    alleles = get_alleles(data)\n",
    "    additive, dominance, AA, Aa, aa = get_genotypic_values(data, alleles)\n",
    "    p, q = get_allele_freqs_wg(data, AA, Aa, aa)\n",
    "    alpha = additive + (dominance*(q-p))\n",
    "    return alpha, AA, aa, p, q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running melezitose\n"
     ]
    }
   ],
   "source": [
    "alpha_vals = {}\n",
    "for p in wgstats:\n",
    "    print \"running %s\" % p\n",
    "    df = pd.DataFrame(index=[\"alpha\", \"p-value\", \"AA\", \"aa\", \"p\", \"q\"])\n",
    "    alpha_vals[p] = df\n",
    "    d = wgstats[p][0]\n",
    "    labels = wgstats[p][1]\n",
    "    for i, locus in enumerate(d):\n",
    "        try:\n",
    "            data = pd.DataFrame(d[locus])\n",
    "            snp = labels[i]\n",
    "            genotypes = [g for g in data.index if \"/\" in g]\n",
    "            data = data.ix[genotypes,:]\n",
    "            pvalue = data['p-value'].dropna()[0]\n",
    "            if len(genotypes) == 3:\n",
    "                alpha, AA, aa, p, q = get_alpha(data)\n",
    "                df[snp] = [alpha, pvalue, AA, aa, p, q]\n",
    "        except Exception as e: \n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X0.10037.01.257</th>\n",
       "      <th>X0.10040.02.394</th>\n",
       "      <th>X0.10044.01.392</th>\n",
       "      <th>X0.10048.01.60</th>\n",
       "      <th>X0.10051.02.166</th>\n",
       "      <th>X0.10054.01.402</th>\n",
       "      <th>X0.10079.02.168</th>\n",
       "      <th>X0.10112.01.169</th>\n",
       "      <th>X0.10113.01.119</th>\n",
       "      <th>X0.10116.01.165</th>\n",
       "      <th>X0.10151.01.86</th>\n",
       "      <th>X0.10162.01.255</th>\n",
       "      <th>X0.1022.02.173</th>\n",
       "      <th>X0.10240.01.410</th>\n",
       "      <th>X0.10262.01.558</th>\n",
       "      <th>...</th>\n",
       "      <th>UMN.CL228Contig1.03.181</th>\n",
       "      <th>UMN.CL22Contig1.02.132</th>\n",
       "      <th>UMN.CL239Contig1.03.104</th>\n",
       "      <th>UMN.CL298Contig1.04.378</th>\n",
       "      <th>UMN.CL306Contig1.04.261</th>\n",
       "      <th>UMN.CL307Contig1.04.143</th>\n",
       "      <th>UMN.CL326Contig1.05.421</th>\n",
       "      <th>UMN.CL339Contig1.05.39</th>\n",
       "      <th>UMN.CL34Contig1.03.89</th>\n",
       "      <th>UMN.CL362Contig1.07.133</th>\n",
       "      <th>UMN.CL363Contig1.01.233</th>\n",
       "      <th>UMN.CL379Contig1.12.117</th>\n",
       "      <th>UMN.CL424Contig1.03.94</th>\n",
       "      <th>UMN.CL54Contig1.07.88</th>\n",
       "      <th>UMN.CL97Contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>alpha</th>\n",
       "      <td>-0.02634913</td>\n",
       "      <td>-0.05690412</td>\n",
       "      <td> 0.07017136</td>\n",
       "      <td> 0.03558499</td>\n",
       "      <td>-0.02413922</td>\n",
       "      <td> 0.04328724</td>\n",
       "      <td>-0.02227026</td>\n",
       "      <td> 0.007385507</td>\n",
       "      <td> 0.01212555</td>\n",
       "      <td>  0.0499099</td>\n",
       "      <td>-0.01000692</td>\n",
       "      <td>-0.03223806</td>\n",
       "      <td>-0.03645684</td>\n",
       "      <td> 0.03023947</td>\n",
       "      <td> 0.003170117</td>\n",
       "      <td>...</td>\n",
       "      <td>  0.0602667</td>\n",
       "      <td>-0.05858812</td>\n",
       "      <td>-0.009311692</td>\n",
       "      <td> 0.02895194</td>\n",
       "      <td> 0.01132639</td>\n",
       "      <td>-0.01049589</td>\n",
       "      <td> 0.01288622</td>\n",
       "      <td>-0.03629308</td>\n",
       "      <td>-0.03593668</td>\n",
       "      <td>-0.01017023</td>\n",
       "      <td>-0.03388197</td>\n",
       "      <td>-0.03779547</td>\n",
       "      <td> 0.008186138</td>\n",
       "      <td> 0.008444086</td>\n",
       "      <td> 0.002275688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p-value</th>\n",
       "      <td>  0.2532518</td>\n",
       "      <td>  0.1788077</td>\n",
       "      <td> 0.05405308</td>\n",
       "      <td>    0.48018</td>\n",
       "      <td>  0.3175836</td>\n",
       "      <td>  0.3643437</td>\n",
       "      <td>  0.4366601</td>\n",
       "      <td>   0.8373771</td>\n",
       "      <td>  0.4662918</td>\n",
       "      <td> 0.07003658</td>\n",
       "      <td>  0.6666358</td>\n",
       "      <td>  0.1382056</td>\n",
       "      <td>  0.3202204</td>\n",
       "      <td>   0.869659</td>\n",
       "      <td>   0.8879599</td>\n",
       "      <td>...</td>\n",
       "      <td>  0.6572657</td>\n",
       "      <td>  0.1048093</td>\n",
       "      <td>   0.7491086</td>\n",
       "      <td>  0.5701891</td>\n",
       "      <td>  0.5970456</td>\n",
       "      <td>  0.2841588</td>\n",
       "      <td> 0.03399721</td>\n",
       "      <td>  0.2289611</td>\n",
       "      <td>  0.2330856</td>\n",
       "      <td>  0.3408673</td>\n",
       "      <td>  0.1710336</td>\n",
       "      <td>   0.629122</td>\n",
       "      <td>   0.8493699</td>\n",
       "      <td>   0.7324662</td>\n",
       "      <td>   0.8618279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA</th>\n",
       "      <td>        G/G</td>\n",
       "      <td>        T/T</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>         A/A</td>\n",
       "      <td>        C/C</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        T/T</td>\n",
       "      <td>        C/C</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>         G/G</td>\n",
       "      <td>...</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>         A/A</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        C/C</td>\n",
       "      <td>        C/C</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>         A/A</td>\n",
       "      <td>         A/A</td>\n",
       "      <td>         G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa</th>\n",
       "      <td>        A/A</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        C/C</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>         G/G</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        C/C</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>         A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>         C/C</td>\n",
       "      <td>        C/C</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        A/A</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>        G/G</td>\n",
       "      <td>         G/G</td>\n",
       "      <td>         G/G</td>\n",
       "      <td>         A/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>  0.7299383</td>\n",
       "      <td>  0.6554878</td>\n",
       "      <td>  0.9542683</td>\n",
       "      <td>  0.8953846</td>\n",
       "      <td>  0.8611111</td>\n",
       "      <td>  0.9832827</td>\n",
       "      <td>  0.9542587</td>\n",
       "      <td>   0.9557927</td>\n",
       "      <td>   0.781155</td>\n",
       "      <td>   0.859375</td>\n",
       "      <td>   0.620743</td>\n",
       "      <td>  0.8929664</td>\n",
       "      <td>  0.5358255</td>\n",
       "      <td>  0.8308824</td>\n",
       "      <td>   0.6168831</td>\n",
       "      <td>...</td>\n",
       "      <td>  0.9097859</td>\n",
       "      <td>  0.8853211</td>\n",
       "      <td>   0.9230769</td>\n",
       "      <td>  0.7602524</td>\n",
       "      <td>  0.9054487</td>\n",
       "      <td>  0.7268519</td>\n",
       "      <td>   0.893617</td>\n",
       "      <td>  0.7092308</td>\n",
       "      <td>  0.7261538</td>\n",
       "      <td>  0.6833856</td>\n",
       "      <td>   0.933642</td>\n",
       "      <td>  0.9095092</td>\n",
       "      <td>   0.5926518</td>\n",
       "      <td>   0.5639344</td>\n",
       "      <td>   0.8323077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>  0.2700617</td>\n",
       "      <td>  0.3445122</td>\n",
       "      <td> 0.04573171</td>\n",
       "      <td>  0.1046154</td>\n",
       "      <td>  0.1388889</td>\n",
       "      <td> 0.01671733</td>\n",
       "      <td> 0.04574132</td>\n",
       "      <td>  0.04420732</td>\n",
       "      <td>   0.218845</td>\n",
       "      <td>   0.140625</td>\n",
       "      <td>   0.379257</td>\n",
       "      <td>  0.1070336</td>\n",
       "      <td>  0.4641745</td>\n",
       "      <td>  0.1691176</td>\n",
       "      <td>   0.3831169</td>\n",
       "      <td>...</td>\n",
       "      <td> 0.09021407</td>\n",
       "      <td>  0.1146789</td>\n",
       "      <td>  0.07692308</td>\n",
       "      <td>  0.2397476</td>\n",
       "      <td> 0.09455128</td>\n",
       "      <td>  0.2731481</td>\n",
       "      <td>   0.106383</td>\n",
       "      <td>  0.2907692</td>\n",
       "      <td>  0.2738462</td>\n",
       "      <td>  0.3166144</td>\n",
       "      <td> 0.06635802</td>\n",
       "      <td>  0.0904908</td>\n",
       "      <td>   0.4073482</td>\n",
       "      <td>   0.4360656</td>\n",
       "      <td>   0.1676923</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 2484 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        X0.10037.01.257 X0.10040.02.394 X0.10044.01.392 X0.10048.01.60  \\\n",
       "alpha       -0.02634913     -0.05690412      0.07017136     0.03558499   \n",
       "p-value       0.2532518       0.1788077      0.05405308        0.48018   \n",
       "AA                  G/G             T/T             G/G            A/A   \n",
       "aa                  A/A             A/A             A/A            C/C   \n",
       "p             0.7299383       0.6554878       0.9542683      0.8953846   \n",
       "q             0.2700617       0.3445122      0.04573171      0.1046154   \n",
       "\n",
       "        X0.10051.02.166 X0.10054.01.402 X0.10079.02.168 X0.10112.01.169  \\\n",
       "alpha       -0.02413922      0.04328724     -0.02227026     0.007385507   \n",
       "p-value       0.3175836       0.3643437       0.4366601       0.8373771   \n",
       "AA                  A/A             G/G             A/A             A/A   \n",
       "aa                  G/G             A/A             G/G             G/G   \n",
       "p             0.8611111       0.9832827       0.9542587       0.9557927   \n",
       "q             0.1388889      0.01671733      0.04574132      0.04420732   \n",
       "\n",
       "        X0.10113.01.119 X0.10116.01.165 X0.10151.01.86 X0.10162.01.255  \\\n",
       "alpha        0.01212555       0.0499099    -0.01000692     -0.03223806   \n",
       "p-value       0.4662918      0.07003658      0.6666358       0.1382056   \n",
       "AA                  C/C             G/G            T/T             C/C   \n",
       "aa                  G/G             C/C            A/A             G/G   \n",
       "p              0.781155        0.859375       0.620743       0.8929664   \n",
       "q              0.218845        0.140625       0.379257       0.1070336   \n",
       "\n",
       "        X0.1022.02.173 X0.10240.01.410 X0.10262.01.558      ...        \\\n",
       "alpha      -0.03645684      0.03023947     0.003170117      ...         \n",
       "p-value      0.3202204        0.869659       0.8879599      ...         \n",
       "AA                 A/A             G/G             G/G      ...         \n",
       "aa                 G/G             A/A             A/A      ...         \n",
       "p            0.5358255       0.8308824       0.6168831      ...         \n",
       "q            0.4641745       0.1691176       0.3831169      ...         \n",
       "\n",
       "        UMN.CL228Contig1.03.181 UMN.CL22Contig1.02.132  \\\n",
       "alpha                 0.0602667            -0.05858812   \n",
       "p-value               0.6572657              0.1048093   \n",
       "AA                          G/G                    A/A   \n",
       "aa                          A/A                    G/G   \n",
       "p                     0.9097859              0.8853211   \n",
       "q                    0.09021407              0.1146789   \n",
       "\n",
       "        UMN.CL239Contig1.03.104 UMN.CL298Contig1.04.378  \\\n",
       "alpha              -0.009311692              0.02895194   \n",
       "p-value               0.7491086               0.5701891   \n",
       "AA                          A/A                     A/A   \n",
       "aa                          C/C                     C/C   \n",
       "p                     0.9230769               0.7602524   \n",
       "q                    0.07692308               0.2397476   \n",
       "\n",
       "        UMN.CL306Contig1.04.261 UMN.CL307Contig1.04.143  \\\n",
       "alpha                0.01132639             -0.01049589   \n",
       "p-value               0.5970456               0.2841588   \n",
       "AA                          G/G                     C/C   \n",
       "aa                          A/A                     A/A   \n",
       "p                     0.9054487               0.7268519   \n",
       "q                    0.09455128               0.2731481   \n",
       "\n",
       "        UMN.CL326Contig1.05.421 UMN.CL339Contig1.05.39 UMN.CL34Contig1.03.89  \\\n",
       "alpha                0.01288622            -0.03629308           -0.03593668   \n",
       "p-value              0.03399721              0.2289611             0.2330856   \n",
       "AA                          C/C                    G/G                   G/G   \n",
       "aa                          A/A                    A/A                   A/A   \n",
       "p                      0.893617              0.7092308             0.7261538   \n",
       "q                      0.106383              0.2907692             0.2738462   \n",
       "\n",
       "        UMN.CL362Contig1.07.133 UMN.CL363Contig1.01.233  \\\n",
       "alpha               -0.01017023             -0.03388197   \n",
       "p-value               0.3408673               0.1710336   \n",
       "AA                          G/G                     A/A   \n",
       "aa                          A/A                     G/G   \n",
       "p                     0.6833856                0.933642   \n",
       "q                     0.3166144              0.06635802   \n",
       "\n",
       "        UMN.CL379Contig1.12.117 UMN.CL424Contig1.03.94 UMN.CL54Contig1.07.88  \\\n",
       "alpha               -0.03779547            0.008186138           0.008444086   \n",
       "p-value                0.629122              0.8493699             0.7324662   \n",
       "AA                          A/A                    A/A                   A/A   \n",
       "aa                          G/G                    G/G                   G/G   \n",
       "p                     0.9095092              0.5926518             0.5639344   \n",
       "q                     0.0904908              0.4073482             0.4360656   \n",
       "\n",
       "        UMN.CL97Contig  \n",
       "alpha      0.002275688  \n",
       "p-value      0.8618279  \n",
       "AA                 G/G  \n",
       "aa                 A/A  \n",
       "p            0.8323077  \n",
       "q            0.1676923  \n",
       "\n",
       "[6 rows x 2484 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alpha_vals[trait_name].head(6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": [
       "iVBORw0KGgoAAAANSUhEUgAAAngAAAHDCAYAAAC3cmcRAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\n",
       "AAALEgAACxIB0t1+/AAAIABJREFUeJzt3X2UpUV9J/BvIzgwAypoYja6OKJQIKKjRqJxfTkSo7hR\n",
       "VkNMomgGzVGMkkhcN9HdRLMqJkfB6CiOEZcxalijBsUY18QsRqMbiCCYgPwEhMSJEXyJwsDIy0zv\n",
       "H/eOtO28dPe93X275vM5h9PcunXrqe7qfvjyPFX1TE1PTwcAgH7ss9wdAABgvAQ8AIDOCHgAAJ0R\n",
       "8AAAOiPgAQB0RsADAOiMgAewC62117TW/n25+wEwXwIewO7ZLBRYcQQ8gN2bWu4OAMzXvsvdAYC5\n",
       "aK1tT/IbSX4qyTOTbE9ybpKXVtW2ndS/MMk/V9WzZpVfneSvq+rFrbWHJfnDJI/O4ErdPyZ5RVV9\n",
       "fjd9OK2q3jKj7NIkX6yqk4ev1yY5c9jmgUn+LsmpVXX18P0fS/JHSZ6Y5O5JrktyRlW9ewE/FoCd\n",
       "cgUPWEl+O8kXk6xL8rIkL0jy8l3U/bMkT2mt3XVHQWvtoUkOS3Jua+0uSf4iya0ZhMZ1SS5P8tHW\n",
       "2oG76cPsW7bTO8paa/sn+VSSn0xyQpLHZHCe/evhe0nyliRHJfnPSVqSDUn+uLX2mD198wBzJeAB\n",
       "K8kVVfWWqvpqVW1K8vEkv7iLuh/K4AraE2eUPTPJv1bVZzK4AvjoJCfVwFczuPJ2zyQPW2D/fiGD\n",
       "APnsqrqoqr6U5LkZBL5nDuscneTzVXVJVX2tqt6R5KeTXLHAYwL8CAEPWElm3zq9LMmhO6tYVf+c\n",
       "5KIMrqTt8IwkHxy+P51B8Hpfa21za+3GJF8Y1jt4gf17RJKvDcPijn7ckOTK3BkaP5Hkha21t7XW\n",
       "ntRa27+qvlBVVusCY2MOHrCSfG/W6y1J7tFae0eSk2aUP6WqPpdBmPutJC9urT0wyYOT/FqStNbu\n",
       "n+SCJJ9J8itJvp7kPkk+Pc8+zVyEcbck92mt3TSrzv4ZhNEkeVWSf0lycpIXJ7m5tfbWJL87DJ0A\n",
       "IxPwgJVkzazXByX59yS/m+SNM8q/Pvz6oSRvbK09MskTklxbVRcN33takv2SPKuqvpckrbV7zqEP\n",
       "s1fVzpyv970km/PDt4V32JIkVbU9yVlJzmqt/UQGgfM1Sb6W5J1zOD7AHgl4wEryuCSvn/H6YUmu\n",
       "rKpvJfnW7MpV9c+ttYsyWNBwXJIPzHj7rklu3xHuhp49/LqrrVFuTHLIjhettXvnh28R/0OSlyS5\n",
       "qaq+OaNeS/LN4UKL/5LkL6vqxqr6RpLXtdZ+IcmDdv1tA8yPgAesJEe11l6e5Pwkj03ylCQv3cNn\n",
       "Ppjk+RmsWH3JjPKLkqxqrZ2W5CMZ3Ka9WwaLLx7VWrtgJ21dkuTZrbXzk9yeQdj8eu4MhB/J4Erc\n",
       "ua21Vyb5dgYLL16fQTj9QgbbsvxCa+11Sb6T5D8lOTLJa+f4MwDYozkFvNbaKUnOSPJ7VXXGsOyw\n",
       "JG9N8oAMbnNckMF+VLe21vbJ4HbJ04dNXJ7kBVX17TH3H9i7vC3JERmEsyR5R5J37eEzH8zgfPTl\n",
       "4arWJElVfaa1dnqSV2Zwi/dDGcyJuy3Jb2YQzn6wBcrQqUn+V5LPZhDkfifJi2a0+f3W2nFJ3pzk\n",
       "rzOYe/elJM+sqr9PktbaUzI4n346yaok1yR5eVX9+Tx+DgC7NTU9vfs5va21szKYY3Jkkv9dVWcO\n",
       "yy9O8rGqek1rbXUGJ6s/r6o/aK29NMmvJnlcVW1trb09yb2q6pcW8XsBOjbcZPhlVfXW5e4LwKSb\n",
       "yzYpZ1fV85LcvKOgtTaV5PQkb0qSqrolgyt4xwyrPC/JxqraOnz95iTPaK0dMK6OAwCwc3u8RVtV\n",
       "l+ykbDrJh3e8bq2tSvLUJG/fUZTkKzM+8tUMwuQRuXOrAAAAFsHIiyyG4e79Sf41yR8Pi9ck2XH1\n",
       "LlW1vbV2a350iwOAOakqG7MDzNFIJ8zW2r0yeO7i9iRPH+7vlAz2ezpgRr27ZDCZeMsoxwMAYM8W\n",
       "fAWvtXZwkr9J8vGqetWsty/PYFHGZ3dUT3JHkppj83ZzBwD2drvak3OP5hPwpmYd6O1J/nYn4S5J\n",
       "NiV5aWvtA0luymAbgnOr6tZ5HO/JSa6bR30m09okn4zx7MXaGM+erI3x7MnaGM+erB3lw7vdJmV4\n",
       "a/XmDK6o3TXJtuE/783g8TrXJZkZ2q6pqp8ffvYNGWzwOZXB7u4vqqrZz2fclen86EINVqYjMrhy\n",
       "azz7YDz7Yjz7Yjz7ckRGGMfdXsGrqm0ZbNS5My/cw2dfmcGVOwAAlpBVaQAAnRHwAAA6I+ABAHRG\n",
       "wAMA6IyABwDQGQEPAKAzAh4AQGcEPACAzgh4AACdEfAAADoj4AEAdEbAAwDojIAHANAZAQ8AoDMC\n",
       "HgBAZwQ8AIDOCHgAAJ0R8AAAOiPgAQB0RsADAOiMgAcA0BkBDwCgMwIeAEBnBDwAgM4IeAAAnRHw\n",
       "AAA6I+ABAHRGwAMA6IyABwDQGQEPAKAzAh4AQGcEPACAzgh4AACdEfAAADoj4AEAdEbAAwDojIAH\n",
       "ANAZAQ8AoDMCHgBAZwQ8AIDOCHgAAJ0R8AAAOiPgAQB0RsADAOiMgAcA0BkBDwCgMwIeAEBnBDwA\n",
       "gM4IeAAAnRHwAAA6I+ABAHRGwAMA6IyABwDQGQEPAKAzAh4AQGcEPACAzgh4AACd2XculVprpyQ5\n",
       "I8nvVdUZw7J7JXl3kqOTbE9yfpJXVNV0a22fJG9M8vRhE5cneUFVfXvM/QcAYJY9BrzW2llJDswg\n",
       "pE3PeGtjks1VdUJrbXWSv01ySpJ3JPn1JI9L8pCq2tpae3uSs5L80pj7v2SmpqZWJzlyzM1eOT09\n",
       "fcuY2wQA9nJzuYJ3dlVd0lq7YEdBa+2gJCdkGHiq6pbW2juTnJxBwHteko1VtXX4kTcnuaK1dsCM\n",
       "spXmyHXHn3bxQYfcdyyN3fSdzbn0E29+RJJLxtIgAMDQHgNeVe0sgBw+fO+aGWVXZXC7Nklakq/M\n",
       "eO+rGcz3OyLJZQvq6QQ46JD75u73fsBydwMAYLfmNAdvJ9YkuW1W2dZh+Y73f3Clrqq2t9ZunfH+\n",
       "XKxdYN8Wxfr16w+95vbxt5lky3hbnThrZ31lZVs76ysr29pZX1nZ1s76ysq2Nj98sWxeFhrwtiRZ\n",
       "NatsTe4MK1uSHLDjjdbaXYb15xNmPrnAvi2KU089NS8789PjbvO8sTY42SZqPBmZ8eyL8eyL8ezH\n",
       "1EI/uNCA95Uk21prh1fVVcOyo3Ln7dfLM5if99nh65bkjiQ1j2M8Ocl1C+zf2G3YsOFByTFjDWQb\n",
       "Nmx4xjnnnHPFONucQGszONlM1HiyYGtjPHuyNsazJ2tjPHuydpQPzyfgTQ3/SVXd3Fr7UJJXJTm5\n",
       "tXaPJC9O8qZh3U1JXtpa+0CSm5K8Msm5VXXrPI53XUa4NDlumzZtOvCxzzlj3G3+yznnnDMx3+Mi\n",
       "uy4TNJ6M7LoYz55cF+PZk+tiPPd6uw14w1urN2ewPcpdk/xMa+11Sf4kyUuSnN1auzrJtgwC3HuS\n",
       "pKre1Vo7LMkXMgiF/5DB1ikAACyy3Qa8qtqWZP/dVDlxN599ZQZX7gAAWEIeVQYA0BkBDwCgMwIe\n",
       "AEBnBDwAgM4IeAAAnRHwAAA6I+ABAHRGwAMA6IyABwDQGQEPAKAzAh4AQGd2+yxaYGWbmppaneTI\n",
       "cbW3fv36Q9/+9rdn9erV42oSgEUg4EHfjlx3/GkXH3TIfcfS2KXXb86VV16Zhz/84WNpD4DFIeBB\n",
       "5w465L65+70fsNzdAGAJmYMHANAZAQ8AoDMCHgBAZ7qegzfmFYRjW4kIK9W2O27Ll7/85WzYsOFB\n",
       "mzZtOnAMTV45PT19yxjaAWCGiQx4LzztdbnoksvfceOWm7eM0s4h9zn6noc+5OceM44VhNdfe/HI\n",
       "bcBKt/XGG/Km99+Qgw455rzHPueMkdq66Tubc+kn3vyIJJeMp3cA7DCRAe/f9jkm//GnjnniqO1c\n",
       "8w/n3TiuFYRbvrN55DagB1blAkw+c/AAADoj4AEAdEbAAwDojIAHANAZAQ8AoDMCHgBAZwQ8AIDO\n",
       "TOQ+eADA/Fx44YX777fffhP3pJkxP1VqB0/B2QMBDwA6sHHjxsMuvf7gSXzSzJHrjj/t4nE8VSrx\n",
       "FJy5EvAAoBOT+qSZSe1Xz8zBAwDojCt4AIxsEeZZmWMFIxDwABiHsc2zMscKRifgATAW5lnB5BDw\n",
       "AJgo2+64LUmOnJqaGmezbvmyVxHwAJgoW2+8IeuOP+39ttWAhRPwAJg4bvfCaAQ8mDBjXo047t3j\n",
       "AVgBBDyYPGNbjXj9tRePoTsArDQCHkygcd2e2vKdzWPoDQArjSdZAAB0xhU8AFgG4376x7HHHnvY\n",
       "qsOPGVdzrHACHgAsj7HNt00Gc27vPZaW6IGABwDLZJzbwZhzy0wCHj9k3LcMTjzxxAe+7GUvy9ln\n",
       "n/2gTZs2HTiGJu1GD7DIxvw0Eds1LQMBj9nGesvgS9denJe+/sM56JBjznvsc84YqS270QMsjXE+\n",
       "TcR2TctDwONHjPuWwYF2pAdYcWzXtLLZJgUAoDMCHgBAZwQ8AIDOCHgAAJ0R8AAAOiPgAQB0RsAD\n",
       "AOiMffCAFW/cT2AZ8tQUfsSYf9c84WGZLcK5Y2LOGwIe0IOxPoHFU1PYjbH9rnnCw0QY23hO2nlj\n",
       "pIDXWntckjcmuVuSO5K8q6re2lq7V5J3Jzk6yfYk5yd5RVVNj9hfgJ0a5xNYYHc84aEvvZ47FjwH\n",
       "r7W2OslHk7y2qo5K8rNJ/kdr7clJNibZXFUPTLIuyeOTnDKG/gIAsAejLLI4NMndk3wySarq+iSX\n",
       "JXlkkhOSnDksvyXJO5OcNFJPAQCYk1EC3lVJvpJhcGutPSDJMUn+Mkmq6ppZdY8e4VgAAMzRggNe\n",
       "VW1LcnKSN7bWvpmkkmxIsibJbbOqbx2WAwCwyBa8yKK19h+SfCzJs6vqr1pr98zg6t0+SVbNqr4m\n",
       "yZYF93KBpvaZmuh9/tavX39oluHnsjvr168/9Jrbl7sXuzaJP7MkufDCC/ffuHHjYeNo69hjjx1L\n",
       "OyvBuMZzMX5vJ/V3bczWzvq6YHvLuWPSv8+9wbY7bsuxxx77hJNPPvnQUdsa9/l2zOeNtRncKV2Q\n",
       "UVbRPibJd6vqr5Kkqr7dWvtYkscluaO1dnhVXTWse1QG8/OW1OoDVh241Mecj1NPPfW85e7DbKee\n",
       "empeduanl7sbuzSJP7Mk2W+//XLp9QdnHEvtv7vvbbn3GPq0EoxrPBfj93ZSf9cWySdHbWBvOXdM\n",
       "+ve5N9h64w2Zvudjzrjm9sk73y7CeWNqoR8cJeBdkeQ+rbWfqqovDFfVPinJ3yb5ZpJXJTm5tXaP\n",
       "JC9O8qYRjrUgt2y9dcsBycSGvA0bNjzjnHPOuWK5+zHThg0bHpQcM7H/YZvEn1ky+LkddMgx59k6\n",
       "YX7GNZ6L8Xs7qb9rY7Y2g3D35CTXjdLQ3nLumPTvc28xqVvVjPm8sXaUDy844FXVFa21FyR5d2tt\n",
       "VQYp81NJTk+yf5KzW2tXJ9mW5Nyqes8oHV2I6e3T25f6mPOxadOmfznnnHMWfPl1MWzatOnAxz7n\n",
       "jOXuxk5tu+O2bHr/ptWbNm0aV2gf247jk/xzm2Tj+htYjJ//JP59LqLrMsKtoGTy/wYm+XeNfkzS\n",
       "eWOkjY6r6k+T/OlO3vp+khNHaRtm23rjDVl3/Gnv73HHcQAYJ48qY0XpdcdxABgnAW+ZbLvjtiQ5\n",
       "cmpqwfMnZ5uYBxwDK8OFF164/3777ZcNGzY8aAxTH8b5wHZgRALeMnG7EVhuGzduPGyw+vuY80ad\n",
       "V3b9tRePqVfAOAh4y8jtRmC5TepqRGA0E70RMAAA8+cKHnulRZgDaf4RABNDwGOvNM45kIn5RwBM\n",
       "FgGPvdY450CafwTAJBHwOjDm241uNQJdcY5kKYx76s/69esPHeWpGAJeB8Z5u9GtRqA3zpEshXFP\n",
       "/bn0+s3J4DGwCyLgdcJWBwC75hzJUpik7c8EPIAVYmpqanXGeIvw2GOPPWzV4ceMqzlgggh4ACvH\n",
       "keuOP+3ica7+vvdYWgImjYAHsIJY/Q3MhSdZAAB0RsADAOiMgAcA0BkBDwCgMwIeAEBnBDwAgM4I\n",
       "eAAAnRHwAAA6I+ABAHRGwAMA6IyABwDQGQEPAKAzAh4AQGcEPACAzgh4AACdEfAAADoj4AEAdGbf\n",
       "5e4AwKTZdsdtSXLk1NTUuJq8cnp6+pZxNQawJwIewCxbb7wh644/7f0HHXLfkdu66Tubc+kn3vyI\n",
       "JJeM3jOAuRHwAHbioEPum7vf+wHL3Q2ABTEHDwCgM67gASyiMc/nO3IcjQD9E/AAFtE45/Ndf+3F\n",
       "Y+gRsDcQ8AAW2bjm8235zuYx9AbYG5iDBwDQGQEPAKAzAh4AQGfMwQOWhdWlAItHwAOWhdWlAItH\n",
       "wAOWjdWlAIvDHDwAgM4IeAAAnRHwAAA6I+ABAHRGwAMA6IyABwDQGQEPAKAzAh4AQGcEPACAzgh4\n",
       "AACdEfAAADoj4AEAdEbAAwDozL6jfLi1dkiSdyb56SS3J9lUVa9trd0rybuTHJ1ke5Lzk7yiqqZH\n",
       "7C8AAHsw6hW8c5J8o6oOzSDk/Wxr7fAkG5NsrqoHJlmX5PFJThnxWAAAzMGCA15r7SeTHJ/kNUlS\n",
       "Vd+qqscn+UaSE5KcOSy/JYOrfCeN2lkAAPZslFu065LckOT5rbXnZnArdmOSi5Kkqq6ZUfeqDG7X\n",
       "AgCwyEa5RXtwkh9P8v2qekiS5yb5gyRPTXLbrLpbk6wZ4VgLMrXPlEUkAMBeZ5QA9N0k00neliRV\n",
       "9Y9JPp7kiUlWzaq7JsmWEY61IKsPWHXgUh8TAGC5jRLwrk6yX5LZIeoLSbYNF1vscFSSy0Y41oLc\n",
       "svXWJQ+VAADLbcEBr6oqyeeSvCpJWmtrM1h08bEkH5pRfo8kL85gxe2Smt4+vX2pjwkAsNxGnaP2\n",
       "3CTHttauy+D27O9U1WeTvCTJQa21q5NcmOTDVfWeEY8FAMAcjLTRcVVdl+S4nZR/N8mJo7QNAMDC\n",
       "WGUKANAZAQ8AoDMCHgBAZwQ8AIDOCHgAAJ0R8AAAOiPgAQB0RsADAOiMgAcA0BkBDwCgMwIeAEBn\n",
       "BDwAgM4IeAAAnRHwAAA6I+ABAHRGwAMA6IyABwDQGQEPAKAzAh4AQGcEPACAzgh4AACdEfAAADoj\n",
       "4AEAdEbAAwDojIAHANAZAQ8AoDMCHgBAZwQ8AIDOCHgAAJ0R8AAAOiPgAQB0RsADAOiMgAcA0BkB\n",
       "DwCgMwIeAEBnBDwAgM4IeAAAnRHwAAA6I+ABAHRGwAMA6IyABwDQGQEPAKAzAh4AQGcEPACAzgh4\n",
       "AACdEfAAADoj4AEAdEbAAwDojIAHANAZAQ8AoDMCHgBAZwQ8AIDOCHgAAJ0R8AAAOiPgAQB0RsAD\n",
       "AOjMvuNopLV2jySXJ/mrqjq5tXavJO9OcnSS7UnOT/KKqpoex/EAANi1cV3Be0uSrUl2BLiNSTZX\n",
       "1QOTrEvy+CSnjOlYAADsxsgBr7X280nun+T9SaZaawcmOSHJmUlSVbckeWeSk0Y9FgAAezZSwGut\n",
       "HZzkzUlOzp1X745Ikqq6ZkbVqzK4XQsAwCIb9QreW5K8bRjmdgS81Ulum1Vva5I1Ix4LAIA5WPAi\n",
       "i9ba05LcL8mvDoumhl9vTrJqVvU1SbYs9FgLNbXPlFXCAMBeZ5RVtM9K8sAkX22tJck9hu09NMkd\n",
       "rbXDq+qqYd2jklw2SkcXYvUBqw5c6mMCACy3BQe8qnruzNettVcnuV9VPb+19v4kr0py8nALlRcn\n",
       "edNIPV2AW7beuuWARMgDAPYqi3UL8yVJDmqtXZ3kwiQfrqr3LNKxdml6+/T2pT4mAMByG8tGx0lS\n",
       "Vb8/49+/m+TEcbUNAMDcWYQAANAZAQ8AoDMCHgBAZwQ8AIDOCHgAAJ0R8AAAOiPgAQB0RsADAOiM\n",
       "gAcA0BkBDwCgMwIeAEBnBDwAgM4IeAAAnRHwAAA6I+ABAHRGwAMA6IyABwDQGQEPAKAzAh4AQGcE\n",
       "PACAzgh4AACdEfAAADoj4AEAdEbAAwDojIAHANAZAQ8AoDMCHgBAZwQ8AIDOCHgAAJ0R8AAAOiPg\n",
       "AQB0RsADAOiMgAcA0BkBDwCgMwIeAEBnBDwAgM4IeAAAnRHwAAA6I+ABAHRGwAMA6IyABwDQGQEP\n",
       "AKAzAh4AQGcEPACAzgh4AACdEfAAADoj4AEAdEbAAwDojIAHANAZAQ8AoDMCHgBAZwQ8AIDOCHgA\n",
       "AJ0R8AAAOiPgAQB0RsADAOiMgAcA0Jl9R/lwa+24JK9Pcvckd0lyVlX9UWvtXkneneToJNuTnJ/k\n",
       "FVU1PWJ/AQDYgwVfwWut/USSjyR5ZVUdleQpSf5na+1RSTYm2VxVD0yyLsnjk5wyhv4CALAHo9yi\n",
       "vSPJSVV1QZJU1VeTXJHk2CQnJDlzWH5LkncmOWm0rgIAMBcLvkVbVd9K8tEdr1trD0jy4CRfHL5/\n",
       "zYzqV2VwuxYAgEU2lkUWrbX7JvlYkj8cFt02q8rWJGvGcaz5mNpnyiISAGCvM3IAaq09PMnnk5xT\n",
       "Va9NsiXJqlnV1gzLl9TqA1YduNTHBABYbiMFvGG4+3iS36yqNw6Lv5JkW2vt8BlVj0py2SjHWohb\n",
       "tt665KESAGC5jbKKdv8kH0zy61V13o7yqro5yYeSvGpY7x5JXpzknNG6On/T26e3L/UxAQCW2yj7\n",
       "4D0jyf2SnN5aO31G+blJXpLk7Nba1Um2JTm3qt4zwrEAAJijUVbRnptBmNuVExfaNgAAC2eVKQBA\n",
       "ZwQ8AIDOCHgAAJ0R8AAAOiPgAQB0RsADAOiMgAcA0BkBDwCgMwIeAEBnBDwAgM4IeAAAnRHwAAA6\n",
       "I+ABAHRGwAMA6IyABwDQGQEPAKAzAh4AQGcEPACAzgh4AACdEfAAADoj4AEAdEbAAwDojIAHANAZ\n",
       "AQ8AoDMCHgBAZwQ8AIDOCHgAAJ0R8AAAOiPgAQB0RsADAOiMgAcA0BkBDwCgMwIeAEBnBDwAgM4I\n",
       "eAAAnRHwAAA6I+ABAHRGwAMA6IyABwDQGQEPAKAzAh4AQGcEPACAzgh4AACdEfAAADoj4AEAdEbA\n",
       "AwDojIAHANAZAQ8AoDMCHgBAZwQ8AIDOCHgAAJ0R8AAAOiPgAQB0RsADAOiMgAcA0BkBDwCgM/su\n",
       "VsOttUcm2ZDknkluT/KGqnrvYh0PAICBRbmC11pbleS8JGdW1eFJnpbkra21By/G8QAAuNNi3aI9\n",
       "Lsl0Vf1ZklTVNUk+nuRXFul4AAAMLVbAOzLJVbPKvpLk6EU6HgAAQ4sV8NYk2Tqr7PvDcgAAFtFi\n",
       "LbK4KckBs8rWJNkylw/v860L86/f+NZXb7/99ttG6cR3v3756lUHHnK3UdrY4ebvXZ/pcTS0CO1N\n",
       "alvjbm9S2xp3e5Pa1rjbm9S2xt3epLY17vYmta1xtzepbY27vUlta9ztTWpbSXLTdzaP9PnFCniX\n",
       "J/mvs8qOSnLZXD780fecPjX2HgEA7CUW6xbtBUnuaK2tT5LW2kOTPCnJ+xbpeAAADE1NT4/zguKd\n",
       "hqHurCQ/lsH8u1dX1XmLcjAAAH5g0QIeAADLw6PKAAA6I+ABAHRGwAMA6IyABwDQGQEPAKAzi7XR\n",
       "8W611h6ZZEOSeya5Pckbquq9O6n3vCS/k2S/JN9O8tKq+sJS9pU9m8d4/kaSF2bwe3dLkv9WVZ9a\n",
       "yr6yZ3Mdzxn1H5Xkc0meX1XvWZpeMlfz+Ptcl2Rj7tza6pVVdf5S9pU9m8d4vijJqRlcyLkxyX+v\n",
       "qr9Zyr4yd621U5KckeT3quqMXdSZVyZa8it4rbVVSc5LcmZVHZ7kaUne2lp78Kx6D0nyliRPG9Y7\n",
       "M8mft9b2W+o+s2vzGM+nJfntJD9XVUcmeUOSD7XW7rrUfWbX5jqeM+rvn+TsJF9LxvqUHsZgHn+f\n",
       "a5L8ZZI3VdUDkrwoyW+21tzlmSDzGM9HZ3COfUpVPSjJq5J8pLV28FL3mT1rrZ2V5GcyeArYTs+j\n",
       "C8lEy/HHe1yS6ar6sySpqmuSfDzJr8yqd1KSvxi+n2H9qSRPWLquMgdzHc+rk/xiVX19+Povktwt\n",
       "yf2WqqPMyVzHc4fXJflokmsz+Ptkssx1PJ+e5Pqq+tCw3t9V1XFVtX1Je8uezHU8H5rkyqraPKz3\n",
       "f5OsSnL/Jewrc3d2VT0vyc27qTPvTLQcAe/IJFfNKvtKkqNnlbVh+UxX7aQey2tO41lVX66qz88o\n",
       "emaSzUm+urjdY57m+veZ1trPZPAfnN8fFrmCN3nmOp4PS3Jda+3s1lq11j7TWnvskvSQ+ZjreP5N\n",
       "kiN2XNlrrZ2Q5N+S/NOi95B5q6pL5lBt3ploOQLemiRbZ5V9f1i+p3pbk6xepH6xMHMdzx9orT0h\n",
       "g0vNJ1fVtsXrGgswp/FsrR2Q5I+TvKCqbluivjF/c/37PDjJE5O8q6paBrfdz2+t3XPxu8g8zGk8\n",
       "q+qqJL+b5IuttRuS/EmSF/lbXdHmnYmWI+DdlOSAWWVrkmyZVbYlP9rxndVjec11PJP8YJLoB5I8\n",
       "y4TfiTTX8Xxdko/M+j9Pt2gnz1zH87tJLqqqC5Okqv4kg4VQj170HjIfcxrP1tpTM5jzfHhV/XiS\n",
       "xyR573AhDSvTvDPRcgS8y5McMavsqCSX7aRe2/GitTaVweXpLy1q75ivuY5nWmsvSPLqJI8fzglh\n",
       "8sx1PJ+Z5LmttWtba9cmeVSSN7XWdrr6i2Uz1/G8KoOreDNNJ7ljkfrFwsx1PJ+a5FNVdV2SVNU/\n",
       "Des8cbE7yKKZdyZajoB3QZI7Wmvrk6S19tAkT0ryvln13pfkqTNWB/1aBv/38pkl6idzM6fxbK09\n",
       "KMkfJDmuqq5c6k4yZ3Maz6q6f1Xdb/j1/kn+PsnLq+rlS91hdmuu59sPZDBn68nDeick2T/J/1u6\n",
       "rjIHcx3Pf0zy+B232FtrhyZZl+TSpesqCzCVXd8JmXcmmpqeXvp50cNfyrNy535Lr66q81prpye5\n",
       "uapeP6yTpHvhAAAAuklEQVT3y0n+R5K7Jvl6kl+vqiuWvMPs1h7Gc0tVnd5ae2eSX85gHGc6rar+\n",
       "z9L2mN2Z69/nrM9ckOSc4a09Jsg8zrc/m+SPMgh2307yW1X1uWXqNrswl/EcXt15XZITk2zP4Grs\n",
       "O6pqw3L1m51rrd0lg9Wz0xlknW3Df96b5FsZIRMtS8ADAGDx2MQSAKAzAh4AQGcEPACAzgh4AACd\n",
       "EfAAADoj4AEAdEbAAwDojIAHANAZAQ8AoDP/HwQxJAgxMpiQAAAAAElFTkSuQmCC\n"
      ],
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fe5734ce150>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(alpha_vals[trait_name].ix['p-value',:], bins=30)\n",
    "plt.title(\"p-values\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": [
       "iVBORw0KGgoAAAANSUhEUgAAAnkAAAHFCAYAAACO6e8yAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\n",
       "AAALEgAACxIB0t1+/AAAIABJREFUeJzt3XuYZFV57/FvcRFhGNHxQhJHaIHhHVAQMaBRuSgqYKJk\n",
       "Ek28BXsCJiCMQhIUzDli4i0ebtFWxIMybQyiiBIxqJiLGnPUKCCoIC/DpcVBBQUFhhkuM9Pnj71b\n",
       "amr6UtVd1dW95vt5nnlqau+19l67Vlf3r/Zae1djdHQUSZIklWWrfjdAkiRJ3WfIkyRJKpAhT5Ik\n",
       "qUCGPEmSpAIZ8iRJkgpkyJMkSSqQIU+SJKlAhjxJkqQCGfI0b0TExoh4cwflRyLi9F62abo6PRap\n",
       "2yLidyLiyxFxQUQ0+t0eTSwi3hARF0bEX/a7LZpftul3A6QOdfIVLaMdlp9tc7lt0xIRjwb+AfgJ\n",
       "sBNAZr59uuXb2V4dUF4HPCczT2hZ9xbg0cBngO2AQeBLmXnFFMfxHGDXzPz0LB9vAK8CHgZ2A67J\n",
       "zA9Ota5efwHwBeAa4Jd1OYCHM3PDOM3ZFvhWZv7dZMc4mWkc/28DK6hOMDwL+HJmntXu9ibr63bW\n",
       "z4H2T1h/srqZeX5EfIXq51dqm2fyJHXTu4F7M/Os+g/UnlOcsZyq/KTrI+I1wBlUfzi3H2f7OwDv\n",
       "AK4Dvgn8bKqAV9uOKhxOpdvH+0ng65n5HuBk4MyI+LM21gHsDXwWuBm4B1hb/+vlGeO2j78OYO8C\n",
       "/j4zTwX+BDglIt7azvam6us2fhb63f6p6nf6syRNyZCnOSMinhoRn4uIX0XE/RHxvYh42STlN0bE\n",
       "iRExHBH3RsSvI+LDEbF1U7GtIuKdEXFnRKyNiE9FxI4z2Of/RMTF4yy/KSI+PIPjeHPLsmsiYmXT\n",
       "8+0j4v31ftZGxFURcUTT+hfXbbu33u+XI2KvifbZC/WZiDcAlzQt/hwTnH2Yqnw728vMT2bm31CF\n",
       "uPGGHEeBvYDnAjtn5vs6OKRJdft4axuBfQAy8z7gV8DvtbEO4EbgYKpjfQ5wCHARcE6Hh9aWTo8f\n",
       "2AN4NrAUIDN/BVwBnNDO9qbq6zZ+Fvra/snqR8R2HbZFaoshT3PJp4EnU/2h2hv4PPCZiBiYpM5b\n",
       "ge8B+wEnAccAf12vawB/RvXH8blUwzjL6nJT7fOpE+zvYuCIiHjU2IKIeAbV8NlF09wmbD502zrU\n",
       "fD7wp/Wx7Qv8O/D5iHhGRDwO+Bfg6/W6g4AH6v3Opn2BHYFbm5b9GNg3Ih47jfKdbG/CP+pZ+XYd\n",
       "jLqp28dLZh6QmUMAEbEAeBJw5VTrImIb4JOZ+d/1sX4HOBQ4JTN7NS2g0+N/CNgZ2L1p2V3A4zvc\n",
       "3lQBrt35hbPd/snqd9oWqS3OydNcsgx4MDN/CRAR/wC8nSq0jExQ5/rMfH/9/1si4ijglcD/qZf9\n",
       "NDPHLr64KSL+BzigjX0+n01/4Y65hGpI6IXAl+tlfwTcnpn/Nc1tTioiFgOvBo7NzLHg9taIOBw4\n",
       "ETiXanjq4swcqesMArtHRGOyP/IR8Xzgf2fm4fXz7ajOhOyXmWs6bOpT6sf7m5bdR/VH97eAX3dY\n",
       "vpPtTXaMf0kV9J8CPJSZ75rqQGgvKHT7eFvLH0s1f/CCcfa9ybrMXM8jP4/UPxs3ZOZP2ziO6ero\n",
       "eDLzx8ATW7ZxANWHk062N1VobTfUzmr7M/OGSep3+rMhtcWQp7nkMcDbI+L3gEU88of2cZPU+WbL\n",
       "82upztqNubJl/a+oPjFPa5+Z+eOI+A5wFI/8UV1GNbF/Jscxmf3rbXytZfnXqY71euA24OKIGAL+\n",
       "LTN/CFzVxraPAlY1PX82sM00Ah5UQXNDZm5sWvZg/fiYaZTvdHvj+RLww8xcCxARX42IX2Xmh8YK\n",
       "1MNsZ7Dp78PfARZGdQHGmFHgHZl5Z5vtb9VW+YjYl+pDxEuo5pfRzrqW4zkhM18+3voumlH/1K/t\n",
       "/vW/GW9vGvra/pb6B8ykLdJEDHmaEyLiMVQh5idUc1NGqM6+rJq4FlBNMG+2hvrKtNq6lvWj1KFr\n",
       "Bvv8DPBXwPERsQfwdKozKzPZ5mTGfsl/PyKal28L3JmZD9Zn5E4FTgHOioik+kP/n1Ns+1CqK/rG\n",
       "HMLmYbJd451tGAvUD0yjfKfb20w9bNnsG1RnPz/UVOYBNg9ThwADmfnxSTbf7eMda8/3qfp6JfC9\n",
       "iHhnZq6cal2To4HvT9Lubpl2/9TDzR8EXpGZOdPtTVPf2t9aPyKWTLct0mQMeZorDqUayjisPgtF\n",
       "RDypjXoLWp4vpDpb18t9XgKcEREH1Nu4tSlMTHebrcODC5v+PxZkDwd+3lJuA0BmrqYKLydGxIHA\n",
       "O4HLIuIp9QTvzdRzfZ4BfLVp8cFUV3GOlVleL5vMdZl5JvBTYOuI2D4zx8L12HH8ZJx6U5XfpsPt\n",
       "bSIitqcKvR/KzLvqxQ1g16nqtqnbx7uJzLwnIi4HPhARF9VhdLx1n8zMB5uqLgc+3OnBdNjXHR9P\n",
       "0362Aj4CvKXlQ8i0tjcDfWn/BPVn+9i1hTDkaa4Yu5DhrqZlr6kfJ5sfdTDVrQfGPBMY+2Q91dyc\n",
       "ae2zacj294HDqC60mMk276Ua1gUgInYGdmlafxXVsTwhM7/ZVG4AuDMidgf2ysx/rdv3nahuy3A1\n",
       "MMDEofdQ4MeZeXe9vW2prtb8y4g4LDP/oz5L1HqmaCLXUp3N2I1qXh/AnsCqCYLmpOUj4t4Ot9fa\n",
       "33sBbwH+DfhWvey3gRvaPJ6pdPt4n0d1Ac2fZ+YX6vUPUN0G5sCI+OwE6x5FPbRXn0n+Xar75HWk\n",
       "w76e8ngmqXc6cMFYwImIYzPzox1urxvz8vrV/s3qAxdMsy3SpLy6VnPFVVTDmn8d1S1I3gC8gOoT\n",
       "7u9GxOMnqLdXRPx1RCyJiD8HjuCRM1EThaqx5VdPsc9FE9SHasj2lVSh6FMdHMd427waeE1EPKue\n",
       "c/XRunwDoJ48fxHVWZsjImIgIv4Y+A7V2bvdgc9FxHH1PoPq3mi/oJqvN5EXsGkAPJrqwoRb2PTi\n",
       "lLZkdcPdi6lelzGvBIbGnkTEYES8r53y7WyvSYPN+/ta4OzM/Fa97x2o5rK9s43DmfLCi6na13ys\n",
       "7ZSnmnT/AFW/jd1X7YVUP1/3TLSu5arhXYGtmYUhvk6Pv152LNX0g0fVP8tHUF0Z30l/j9fXU66P\n",
       "iNd30h+9aP9E9eu5eO3+rEttm/RMXn2m4BYeOTMy5nlUAfFjwNOo/qhdRn25fn06+gxgbOLvdcAx\n",
       "TUMm0iYy89aIeBPwNuAvqO4f9Xqqe1C9jepT7nifzj9I9Yl3bLj0w1S3G2GC8r+5NUlm3jLJPv+W\n",
       "KgC9aYImf4bqZ/xH9Typdo5jom2uoPok/w2qoZlTgdavLzoWeG9d7gl1uTMy8wyAiDie6ga5Z1Pd\n",
       "APe7wBEtw3itXgjcExGn1nWuBP6rfj7VXL6JnEIVRk+lmht5ZTZ9KwPV8PDBHZSfdH1ELAP+gPp3\n",
       "TdTf+pCZl2bmhqi+CuoMqjNdi4GTMvPS5gbXFyqcw6a/D38b2LGe6zhmFPhfTRdeTNW+1mOdtHxm\n",
       "XjP2oSAiXkp1NvffgNPreZfjrmvZ/lqqIf0RZkfbxx8RS6muBN+GTW/Q/Il2tjdZX7ezniqMHR0R\n",
       "fzd2Ic4st3+q+lO9F6SONUZHJz6rPRbyMnOzM34RcQlwR2aeUH9C/jrVKegPR8SJVH/YDs7MdRHx\n",
       "Iaqhpj/tyVFoixQRG6n+aH+g322ZbyLiicDtwBMzs/XiFW0BImJXYDBn8LVm801Ut1j6ambe2++2\n",
       "dGpL7C/N3LSGayNiIdWtF84GqD8VfYTqZrNQDfuc1zSB9BxgWT0RWlL/vQC42oCnLczi+RjwpOlq\n",
       "68KLiPgnqgntDwDvp57nk5k3NxVbRTV0CxBUX7Ez5haqQLkn1TwZSf21F5ve208qWj303s69I+eq\n",
       "dr/JQ/qNqULefVTz7oYy8/v11V9fAY6k+oqWZut45HYWC2i6P1lmboyIB9n8dhfStI03jUDtcchH\n",
       "VL/Dn1PPXTsme/f1Z3NCZv53v9swXfV8zEN55Ns1pLZMGvLqCyXe0PT8/0XEZVSTfbdrKb6A6ka0\n",
       "1I+/GZqN6gvjt2taL0nqo8z8GdUHds1xmXk+j1xQJrVtqqtrH0d1wUTz3fq3phpyfX5ELGlatxeP\n",
       "DMVeByyluloQquHb9Wx+le5Eiv5EKUmS1KZpD9VPNVz7XOCCiDggM2+LiKdT3YfsMGBnqttELI/q\n",
       "zvnHA2N3Qh+muvP+p6mGfE8DLpridg6tDmf2bgOgTQ1Q3frDPuifAeyDfhrA17/fBrAP+m0A+6Df\n",
       "BmZSearh2ssj4t3AVyJilOrCi2Mz87sRcQLw0Yi4ieqrlS7K+nseM/P8iNiN6r5bDap7dr2xw7aN\n",
       "sOnFG5p9I9gH/TaCfdBPI/j699sI9kG/jWAfzEuT3ievj0bZ/ApdzZ49qYbW7YP+sQ/6a8+1a9fm\n",
       "CSecsGx4ePi2DuveMDo6unbqYpqC74H+sw/6b09m8Nr73bWSNI4bbriBa+543KUHvfastuvcd/dq\n",
       "rvnSOc+i+qo6SeorQ54kTWDhosXstPPu/W6GJE2L9xmTJEkqkCFPkiSpQIY8SZKkAhnyJEmSCmTI\n",
       "kyRJKpAhT5IkqUCGPEmSpAIZ8iRJkgpkyJMkSSqQIU+SJKlAhjxJkqQCGfIkSZIKZMiTJEkqkCFP\n",
       "kiSpQIY8SZKkAhnyJEmSCmTIkyRJKpAhT5IkqUCGPEmSpAIZ8iRJkgpkyJMkSSqQIU+SJKlAhjxJ\n",
       "kqQCGfIkSZIKZMiTJEkqkCFPkiSpQNv0uwGS1GuNRmMHYGm75QcHB3d50Yte1MMWSVLvGfIkbQmW\n",
       "7nfkyVctXLS4rcI3PwzfPPeL7PzUZ/W4WZLUO4Y8SVuEhYsWs9POu7ddfs3dq3vYGknqPefkSZIk\n",
       "FciQJ0mSVCBDniRJUoEMeZIkSQUy5EmSJBXIkCdJklQgQ54kSVKBDHmSJEkFMuRJkiQVyJAnSZJU\n",
       "IEOeJElSgQx5kiRJBTLkSZIkFciQJ0mSVCBDniRJUoEMeZIkSQUy5EmSJBXIkCdJklQgQ54kSVKB\n",
       "DHmSJEkFMuRJkiQVyJAnSZJUIEOeJElSgQx5kiRJBTLkSZIkFciQJ0mSVCBDniRJUoEMeZIkSQUy\n",
       "5EmSJBXIkCdJklQgQ54kSVKBDHmSJEkFMuRJkiQVyJAnSZJUIEOeJElSgQx5kiRJBTLkSZIkFciQ\n",
       "J0mSVCBDniRJUoG2abdgRDwWuA74SmYuj4gnAB8DngZsBC4DTsnM0YjYCjgDeHld/TrgmMy8q6ut\n",
       "lyRJ0rg6OZP3fmAdMFo/Pw9YnZl7APsBhwDH1eveCBwM7JuZS4DbgXO70mJJkiRNqa2QFxF/ADwV\n",
       "uBBoRMSOwFHA2QCZuRb4CPC6usrRwHmZua5+fg6wLCK272LbJUmSNIEpQ15EPI4qpC3nkbN4ewJk\n",
       "5s1NRVdRDd0CBHBj07pb6n3tOcP2SpIkqQ3tnMl7P/DBOtCNhbwdgIdayq0DFtT/X1A/ByAzNwIP\n",
       "Nq2XJElSD0164UVEvAzYFXh9vahRP94PbNdSfAGwpv7/GuA3Q7MRsXVdfg3tG+igrLproOVRs2+g\n",
       "5VEzMDg4uMvND8/evujsd53GN9DyqNk30PKo2TfApiOjHZnq6to/AfYAbokIgMfWdZ4BrI+IJZm5\n",
       "qi67F3Bt/f/rgKXAN+rnAawHsoO2XdFBWfWGfdB/9kEXrFixgpPO/tps7evSWdnRlsP3QP/ZB/3V\n",
       "mLrI+CYNeZn5Z83PI+J0YNfM/POIuBB4G7C8vr3K8cCZddFh4MSI+DRwH3AacFFmPthB2w4HRjoo\n",
       "r+4ZoHpT2wf9M4B90DVDQ0N7wz6zEr6GhoaWrVy58vrZ2FfhBvA90G8D2Af9NjCTym3fJ28cJwAf\n",
       "jYibgA1UIe7jAJl5fkTsBlxJlUC/S3VblU6MMINTlOqKEeyDfhvBPpix4eHhHQ967Vmzta/bVq5c\n",
       "aZ91zwi+B/ptBPtgXuoo5GXm3zX9/9fAKyYpexrVGTxJkiTNMr/WTJIkqUCGPEmSpAIZ8iRJkgpk\n",
       "yJMkSSqQIU+SJKlAhjxJkqQCGfIkSZIKZMiTJEkqkCFPkiSpQIY8SZKkAhnyJEmSCmTIkyRJKpAh\n",
       "T5IkqUCGPEmSpAIZ8iRJkgpkyJMkSSqQIU+SJKlAhjxJkqQCGfIkSZIKZMiTJEkqkCFPkiSpQIY8\n",
       "SZKkAhnyJEmSCmTIkyRJKpAhT5IkqUCGPEmSpAIZ8iRJkgpkyJMkSSqQIU+SJKlAhjxJkqQCGfIk\n",
       "SZIKZMiTJEkqkCFPkiSpQIY8SZKkAm3T7wZIUicajcYOwNIOq3VaXpLmPUOepPlm6X5HnnzVwkWL\n",
       "265wx61X9bA5kjQ3GfIkzTsLFy1mp513b7v8mrtX97A1kjQ3OSdPkiSpQIY8SZKkAhnyJEmSCmTI\n",
       "kyRJKpAhT5IkqUCGPEmSpAIZ8iRJkgpkyJMkSSqQIU+SJKlAhjxJkqQCGfIkSZIKZMiTJEkqkCFP\n",
       "kiSpQIY8SZKkAhnyJEmSCmTIkyRJKpAhT5IkqUCGPEmSpAIZ8iRJkgpkyJMkSSqQIU+SJKlAhjxJ\n",
       "kqQCGfIkSZIKZMiTJEkqkCFPkiSpQNv0uwGSVIoN6x8CWNpoNDqtesPo6Oja7rdI0pbMkCdJXbLu\n",
       "3jvZ78iTL1y4aHHbde67ezXXfOmcZwFX965lkrZEhjxJ6qKFixaz086797sZkuScPEmSpBIZ8iRJ\n",
       "kgpkyJMkSSqQIU+SJKlAhjxJkqQCTXl1bUQcAbwT2BEYBc7LzA9ExBOAjwFPAzYClwGnZOZoRGwF\n",
       "nAG8vN7MdcAxmXlXD45BkiRJLSY9kxcRvwV8BnhTZu4F/D7wzoh4PnAesDoz9wD2Aw4BjqurvhE4\n",
       "GNg3M5cAtwPn9uYQJEmS1Gqq4dqNwGsy81sAmXkrsIoq1B0FnF0vXwt8BHhdXe9oqjN+6+rn5wDL\n",
       "ImL77jZfkiRJ45l0uDYz7wS+MPY8Il4I7AJ8s15/c1PxVVRDtwAB3Ni07haqQLkncO2MWy1JkqRJ\n",
       "tfWNFxHxUqrh2QXA8fXjQy3F1tXLqR/HzuKRmRsj4sGm9e0Y6KCsumug5VGzb6DlUbXBwcFdbn64\n",
       "363orsHBwV2ANf1uxxwz0PKo2TfQ8qjZN8CmJ8060lbIy8wvArtExFKqM3sfB7ZrKbaAR35JrQF+\n",
       "MzQbEVvX5Tv5JXZFB2XVG/ZB/9kHLVasWMFJZ3+t383oqhUrVlza7zbMYb4H+s8+6K/GdCtOGvIi\n",
       "Yk9gSWZeDpCZN0TEZcD+wIaIWJKZq+rie/HIUOx1wFLgG2ObAtYD2UHbDgdGOiiv7hmgelPbB/0z\n",
       "gH0wrqGhob1hn6JC0dDQ0LKVK1de3+92zDED+B7otwHsg34bmEnlqc7kLQIuiojnZeYPIuKxwIuo\n",
       "zuStA94GLK+XHw+cWdcbBk6MiE8D9wGnARdl5oMdtG2EGZyiVFeMYB/02wj2wSaGh4d3POi1Z/W7\n",
       "GV01PDx828qVK+3n8Y3ge6DfRrAP5qVJr67NzG8DJwCfiYhVwJXAV4B/rJcvjIibgP8BPpuZH6/r\n",
       "nQ98sS5/I7A18KZeHYQkSZI2NeWcvMz8BPCJcVb9GnjFJPVOozqDJ0mSpFnm15pJkiQVyJAnSZJU\n",
       "IEOeJElSgQx5kiRJBTLkSZIkFciQJ0mSVCBDniRJUoEMeZIkSQUy5EmSJBXIkCdJklQgQ54kSVKB\n",
       "DHmSJEkFMuRJkiQVyJAnSZJUIEOeJElSgQx5kiRJBTLkSZIkFciQJ0mSVCBDniRJUoEMeZIkSQUy\n",
       "5EmSJBXIkCdJklQgQ54kSVKBDHmSJEkFMuRJkiQVyJAnSZJUIEOeJElSgQx5kiRJBTLkSZIkFciQ\n",
       "J0mSVCBDniRJUoEMeZIkSQUy5EmSJBXIkCdJklQgQ54kSVKBDHmSJEkFMuRJkiQVyJAnSZJUIEOe\n",
       "JElSgQx5kiRJBTLkSZIkFciQJ0mSVCBDniRJUoEMeZIkSQUy5EmSJBXIkCdJklQgQ54kSVKBDHmS\n",
       "JEkFMuRJkiQVyJAnSZJUIEOeJElSgQx5kiRJBdqm3w2QtOVqNBo7AEs7rNZpeUnaIhnyJPXT0v2O\n",
       "PPmqhYsWt13hjluv6mFzJKkchjxJfbVw0WJ22nn3tsuvuXt1D1sjSeVwTp4kSVKBDHmSJEkFMuRJ\n",
       "kiQVyJAnSZJUIEOeJElSgQx5kiRJBTLkSZIkFciQJ0mSVCBDniRJUoEMeZIkSQUy5EmSJBXIkCdJ\n",
       "klQgQ54kSVKBDHmSJEkF2qadQhFxGPBuYCdga+DczPzHiHgC8DHgacBG4DLglMwcjYitgDOAl9eb\n",
       "uQ44JjPv6vIxSJIkqcWUZ/Ii4reAfwFOy8y9gCOAv4+I5wDnAaszcw9gP+AQ4Li66huBg4F9M3MJ\n",
       "cDtwbvcPQZIkSa3aGa5dD7wuM78KkJm3ANcDBwJHAWfXy9cCHwFeV9c7GjgvM9fVz88BlkXE9t1r\n",
       "viRJksYz5XBtZv4S+PzY84jYHXg68L16/c1NxVdRDd0CBHBj07pbqELlnsC1M2q1JEmSJtXRhRcR\n",
       "sRj4AvC+etFDLUXWAQvq/y+onwOQmRuBB5vWS5IkqUfauvACICL2p5qbN5SZZ0TEM4HtWootANbU\n",
       "/18DbN9Uf+u6/BraM9Bu29R1Ay2Pmn0DLY9FGhwc3OXmh/vdiv4bHBzchfZ/N24pBloeNfsGWh41\n",
       "+wbYdFS0I+1eXbs/cDnwxsy8tF58I7AhIpZk5qp62V48MhR7HbAU+MbYZqjm92WbbbuizXLqHfug\n",
       "/4rugxUrVnDS2V/rdzP6bsWKFZdOXWqLVfR7YJ6wD/qrMd2KU4a8iHg08Bk2DXhk5v0RcQnwNmB5\n",
       "RDwWOB44sy4yDJwYEZ8G7gNOAy7KzAfbbNvhwEibZdVdA1RvavugfwbYAvpgaGhob9hniw84Q0ND\n",
       "y1auXHl9v9sxxwywBbwH5rgB7IN+G5hJ5XbO5C0DdgXeExHvaVp+EXAC8NGIuAnYQBXiPg6QmedH\n",
       "xG7AlVQp9LtUt1Vp1wgzOEWprhjBPui3EQrug+Hh4R0Peu1Z/W5G3w0PD9+2cuXKYvt5hkYo+D0w\n",
       "T4xgH8xL7VxdexFVoJvIKyapexrVGTxJkiTNIr/WTJIkqUCGPEmSpAK1fQsVSVL3bVj/EMDSRqPj\n",
       "C+huGB0dXdv9FkkqhSFPkvpo3b13st+RJ1+4cNHituvcd/dqrvnSOc8Cru5dyyTNd4Y8SeqzhYsW\n",
       "s9POu/e7GZIK45w8SZKkAhnyJEmSCmTIkyRJKpAhT5IkqUCGPEmSpAIZ8iRJkgpkyJMkSSqQIU+S\n",
       "JKlAhjxJkqQCGfIkSZIKZMiTJEkqkCFPkiSpQIY8SZKkAhnyJEmSCmTIkyRJKpAhT5IkqUCGPEmS\n",
       "pAIZ8iRJkgpkyJMkSSqQIU+SJKlAhjxJkqQCGfIkSZIKZMiTJEkqkCFPkiSpQIY8SZKkAhnyJEmS\n",
       "CmTIkyRJKpAhT5IkqUCGPEmSpAIZ8iRJkgpkyJMkSSqQIU+SJKlAhjxJkqQCGfIkSZIKZMiTJEkq\n",
       "kCFPkiSpQIY8SZKkAhnyJEmSCmTIkyRJKpAhT5IkqUCGPEmSpAIZ8iRJkgpkyJMkSSqQIU+SJKlA\n",
       "hjxJkqQCGfIkSZIKZMiTJEkqkCFPkiSpQIY8SZKkAhnyJEmSCmTIkyRJKpAhT5IkqUCGPEmSpAIZ\n",
       "8iRJkgpkyJMkSSqQIU+SJKlAhjxJkqQCGfIkSZIKZMiTJEkqkCFPkiSpQNv0uwGSytBoNHYAlnZY\n",
       "rdPykqQ2GfIkdcvS/Y48+aqFixa3XeGOW6/qYXMkactmyJPUNQsXLWannXdvu/yau1f3sDWStGVz\n",
       "Tp4kSVKBDHmSJEkFanu4NiKOA84C3p6ZZ9XLngB8DHgasBG4DDglM0cjYivgDODl9SauA47JzLu6\n",
       "2H5JkiSNo60zeRFxLvBcqqA22rTqPGB1Zu4B7AccAhxXr3sjcDCwb2YuAW4Hzu1SuyVJkjSJdodr\n",
       "P5qZRwP3jy2IiIXAUcDZAJm5FvgI8Lq6yNHAeZm5rn5+DrAsIrbvRsMlSZI0sbZCXmZePc7iJfW6\n",
       "m5uWraIaugUI4MamdbfU+9uz82ZKkiSpEzO58GIB8FDLsnX18rH1Y2fxyMyNwINN6yVJktQjM7lP\n",
       "3hpgu5ZlC+rlY+t/MzQbEVvX5dfQnoEZtE0zM9DyqNk30PI45w0ODu5y88P9bsWWYcP6hzjwwAMP\n",
       "Xb58+S6d1DvuuONuefazn/1Ar9rVZQMtj5p9Ay2Pmn0DbDoq2pGZhLwbgQ0RsSQzV9XL9gKurf9/\n",
       "HdVXFn2jfh7AeiDb3P4VM2ibusM+6L950wcrVqzgpLO/1u9mbBHW3Xsno49/3lk3P9z+t4vcd/dq\n",
       "tt122x62qmfmzXugYPZBfzWmW7HTkNcY21lm3h8RlwBvA5ZHxGOB44Ez67LDwIkR8WngPuA04KLM\n",
       "fLDNfR0OjHTYPnXHANWb2j7onwHmWR8MDQ3tDftc2u92bCk6/XYRgKGhoWUrV668vkdN6rYB5tl7\n",
       "oEAD2Af9NjCTylOGvHqY9X6qW6c8CnhuRLwL+CfgBOCjEXETsIEqxH0cIDPPj4jdgCupguF3qW6r\n",
       "0q4RZnCKUl0xgn3QbyPMkz4YHh7e8aDXntXvZmgSw8PDt61cuXJe/Dw1GWGevAcKNoJ9MC9NGfIy\n",
       "cwPw6EmKvGKSuqdRncGTJEnSLPJrzSRJkgpkyJMkSSqQIU+SJKlAhjxJkqQCGfIkSZIKZMiTJEkq\n",
       "kCFPkiSpQIY8SZKkAhnyJEmSCmTIkyRJKpAhT5IkqUCGPEmSpAIZ8iRJkgpkyJMkSSqQIU+SJKlA\n",
       "hjxJkqQCGfIkSZIKZMiTJEkqkCFPkiSpQIY8SZKkAhnyJEmSCmTIkyRJKpAhT5IkqUCGPEmSpAIZ\n",
       "8iRJkgpkyJMkSSqQIU+SJKlAhjxJkqQCGfIkSZIKZMiTJEkqkCFPkiSpQIY8SZKkAhnyJEmSCmTI\n",
       "kyRJKpDBtmHcAAAJf0lEQVQhT5IkqUCGPEmSpAJt0+8GSJp7Go3GDsDSDqt1Wl6S1EOGPEnjWbrf\n",
       "kSdftXDR4rYr3HHrVT1sjiSpU4Y8SeNauGgxO+28e9vl19y9uoetkSR1yjl5kiRJBTLkSZIkFciQ\n",
       "J0mSVCDn5EnSFmDD+ocAljYajU6r3jA6Orq2+y2S1GuGPEnaAqy79072O/LkCzu5Yvq+u1dzzZfO\n",
       "eRZwde9aJqlXDHmStIXo9IppSfObc/IkSZIKZMiTJEkqkCFPkiSpQIY8SZKkAhnyJEmSCmTIkyRJ\n",
       "KpC3UJEK12g0dgCWdlit0/IqkDdQluY3Q55UvqX7HXnyVZ3cBPeOW6/qYXM0X3gDZWl+M+RJW4BO\n",
       "b4K75u7VPWyN5hNvoCzNX87JkyRJKpAhT5IkqUCGPEmSpAIZ8iRJkgrkhRfSPOLtUCRJ7TLkSfOL\n",
       "t0ORJLXFkCfNM94ORZLUDufkSZIkFciQJ0mSVCBDniSprxqNxg6NRmP/5n/Lly/f++qrr2b58uV7\n",
       "t65r+rdDv9suzWXOyZP6ZLIrZQcHB3dZsWIFQ0NDew8PD+/YtMorZVWizS4ouvlhOOnsrwH7XHrQ\n",
       "a8/arILfkStNzZAndcF0b20y0Ze/T/QHzitlVSq/I1fqPkOe1B3TurWJV8pKknrFkCd1iYFNkjSX\n",
       "GPIkSV2zYf1DAEsbjUYn1ZxrKvWAIU+S1DXr7r2TieaaTsS5plJv9DTkRcQBwBDweOBh4L2Z+Yle\n",
       "7lOS1F9OXZDmhp7dJy8itgMuBc7OzCXAy4APRMTTe7VPSZIkVXp5Ju8wYDQzLwbIzJsj4nLg1cDf\n",
       "9nC/kqTCTXPuH8ANo6Oja7vfImnu6WXIWwqsall2I7B/D/epgk3zXnTgL3WpONOZ+3fPL27l+1/5\n",
       "4GsbjcYNHe7O3yGal3oZ8hYA61qWPVAv1xz16B0XPfk5zz3o5ce9YZDzzh9++feu+cHPp6qzccP6\n",
       "DWvu/sklo6OjD/e4eR3fi26av9QfXT8+0EnbOigrqQumM/ev02A4W9+sMc0PsdP5XdV2nZZv3llt\n",
       "0J1/ehny7gO2b1m2AFjTZv2BrrZGbXn2c5531M/vGX3z/73wS/xizTZv3nlgnymD25q7b2+85NAD\n",
       "WL58eWuo76oDDzxwt4c6rPPAmrtZ8uxXXrj9Y57Udp1f/Sx59ILH0Wmd+zqcPH7/PXcw2lEN61jH\n",
       "OjOts8NOO3dYCw488MBDly9fvkvHFTvbx26/aux61mz8rmq3ztW3w/JTL2DdvfdfOjg4uAy4vu0d\n",
       "qVsGqEZBp6UxOtrp26Q9EfFiYGVmLm5adjFwfWa+oyc7lSRJEtDDq2uBrwLrI2IQICKeAbwY+Oce\n",
       "7lOSJEn08Ewe/CbYnQs8kWr8//TMvLRnO5QkSRLQ45AnSZKk/ujlcK0kSZL6xJAnSZJUIEOeJElS\n",
       "gQx5kiRJBTLkSZIkFaiX33jRtoh4C3AMVei8DXhDZt4yTrknUt2S5RnAKPA94MTM/OUsNrdIHfbB\n",
       "+4FnAttSfdXP8Zl51yw2tzjtvv512acDFwJbZeY+s9fK8kTEAcAQ8HjgYeC9mfmJccodDZxK9TN/\n",
       "F9XvnStns62l6qAPtgfOBI4Hfjcze/o1Y1uKDl7/NwF/QZUb1gJvycx/n822lqqDPvgr4FigQfXt\n",
       "Yadm5n9Mtu2+n8mLiD8ATgCel5lLgCuAiyYofj7w88zcE9gb2A74h1lpaME67IPzqL6TeG+q71l8\n",
       "FPDu2WhnqTp5/SPiEOCTwNdnr4VliojtgEuBs+vX/WXAB+oQ3VxuX6oPNi+ry50NfC4itp3tNpem\n",
       "3T6ofRu4dTbbV7oO3gMvA94KvCQzlwLvBS6JiEfNdptL00Ef/CGwAjgkM/cC3gd8tv7wM6G+hzzg\n",
       "aOCfms7GfRB4ZkTsMU7ZlcDfA2TmBuA/gafNSivL1kkfXAD8bWaOZuZ6qkCy7yy1s1SdvP4/A36P\n",
       "Hn9Z+hbiMGA0My8GyMybgcuBV7eUex3wr/V66vIN4NDZa2qx2u0DgOWZeeZsNm4L0O7rfxPwysz8\n",
       "af38X4HHALvOVkML1kkfvDozf1E/v4KqD54y2cbnQsgLmr58NzPXAqsZJ7xl5ufHDrAeNnwV8NlZ\n",
       "amfJOumDyzPz5wAR0QD+EPivWWpnqTp5/W/MzPtnsW0lWwqsall2I5u/7pv0T23VOOXUuXb7AIdn\n",
       "e6Kt1z8zf5SZ32xa9EdUv6PGnVKijrTbBz/MzG8DRMTWVKM/36cKfxOalTl5EfEqqvHmVvfUj+ta\n",
       "lq8DFkyyvR8AS6iGtN7fjTaWrgd90AD+EXgS8K5utLFk3X791RUL2Px1f4DNX/fxyq0DduhRu7Yk\n",
       "7faBeqPj1z8iDqX6u/un9YiaZqajPoiIdwAnAndQndnbONnGZyXkZeangE+Nty4irgFax5QXUE0q\n",
       "nGh7+0TEjlSTcC8HXtKlpharm30QETsAn6CaJPqCzJywr1Tp9ntAXXEf7b3ua9g80Nk/3dFuH6g3\n",
       "Onr96wuQzgD+JDP/s8dt21J01AeZ+Q7gHRHxUuBrEXFgZk54Nm8uDNdeR3W6EoCIWAg8GfhBc6GI\n",
       "2CYiButwRx0szgVe5AToGWurD+p12wGfp/oBfFFm/nq2Glmwtl9/ddV1wJ4ty/YCrh2nXIw9qc9i\n",
       "L6UaKtHMtNsH6o22X/+IOAY4nWrivwGve9rqg4g4JCL2H3uemV+kuhDpBZNtfC6EvGHg9RHx5Pr5\n",
       "qcB/Z+YmV1HVk/xPAf6mafEfAz/MzIdno6EFG6aNPqidDtwPDNZ9opkbpv3XX93zVWB9RAwCRMQz\n",
       "gBcD/9xS7p+BlzZd7XYs1adv56LOXLt9QL2+Uf+3Md56dayt1z8i9qa6k8VhmXnDbDeycO2+Bw4E\n",
       "LoiIx9blnk4VDr832cYbo6Oj3W5wxyLiJOA4qtB5I/AXY1fxRMSPgD/KzB9FxJ7AB6g+RT9MNenz\n",
       "TZmZ/Wl5OTrogweAO6mC3pgHMvOZs93mknTw+l9AddXV1nXZh6muzHJ+2DTUv1DPBZ5INQ/m9My8\n",
       "NCLeA9yfme+uy70K+F9Utwz6KfDGzLy+T80uSjt9EBEvAL5YV3kU9c89cGxmXtiPdpdiitd/TWa+\n",
       "JyI+QnWh409bqp+cmV+e3RaXp833wNZUtyt7FfAgsB44IzOHJ9v2nAh5kiRJ6q65MFwrSZKkLjPk\n",
       "SZIkFciQJ0mSVCBDniRJUoEMeZIkSQUy5EmSJBXIkCdJklQgQ54kSVKBDHmSJEkF+v+l/hi4ERHb\n",
       "2QAAAABJRU5ErkJggg==\n"
      ],
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fe5739b5610>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(alpha_vals[trait_name].ix['alpha',:], bins=30)\n",
    "plt.title(\"alpha values $\\mu %.4f \\pm %.4f \\ [%.4f, %.4f]$\" % (np.mean(alpha_vals[trait_name].ix['alpha',:]),\n",
    "                                                            np.std(alpha_vals[trait_name].ix['alpha',:]),\n",
    "                                                            np.min(alpha_vals[trait_name].ix['alpha',:]),\n",
    "                                                             np.max(alpha_vals[trait_name].ix['alpha',:])))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def is_homozygous(gt):\n",
    "    if len(set([x.strip() for x in gt.split(\"/\")])) == 1:\n",
    "        return True\n",
    "    return False\n",
    "\n",
    "def get_allele_counts(counts):\n",
    "    a = {}\n",
    "    het = 0\n",
    "    for gt in counts.index:\n",
    "        for allele in [x.strip() for x in gt.split(\"/\")]:\n",
    "            if not allele in a:\n",
    "                a[allele] = 0\n",
    "            a[allele] += counts[gt]\n",
    "        if not is_homozygous(gt):\n",
    "            het += counts[gt]\n",
    "    return sorted(a.items(), key=lambda x: x[1], reverse=True), het\n",
    "\n",
    "def get_correction(n):\n",
    "    #for finite sample size\n",
    "    return (2*n)/(2*n-1)\n",
    "\n",
    "def get_allele_freqs(locus):\n",
    "    locus = locus[locus != '?/?']\n",
    "    locus = locus[locus != 'NA']\n",
    "    c = locus.value_counts()\n",
    "    c = c.sort(inplace=False, ascending=False)\n",
    "    allele_counts = get_allele_counts(c)\n",
    "    total_alleles = 2.0*sum(c)\n",
    "    num_individuals = sum(c)\n",
    "    A = \"\"\n",
    "    a = \"\"\n",
    "    P = 0\n",
    "    Q = 0\n",
    "    if len(allele_counts[0]) == 2:\n",
    "        A = allele_counts[0][0][0]\n",
    "        a = allele_counts[0][1][0]\n",
    "        P = allele_counts[0][0][1]\n",
    "        Q = allele_counts[0][1][1]\n",
    "    else:\n",
    "        A = allele_counts[0][0][0]\n",
    "        P = P = allele_counts[0][0][1]\n",
    "    PQ = allele_counts[-1]\n",
    "    p = P/total_alleles\n",
    "    q = Q/total_alleles\n",
    "    assert p + q == 1.0\n",
    "    He = 2 * p * q * get_correction(num_individuals)\n",
    "    Ho = PQ*1.0/num_individuals\n",
    "    Fis = 1 - (Ho/He)\n",
    "    #print p, q, He, Ho, Fis\n",
    "    ret = pd.Series({\"p\":p, \n",
    "                      \"q\":q,\n",
    "                      \"P\":P,\n",
    "                      \"Q\":Q,\n",
    "                      \"He\":He,\n",
    "                      \"Ho\":Ho, \n",
    "                      \"Fis\":Fis,\n",
    "                    \"PQ\": PQ,\n",
    "                    \"total_alleles\":total_alleles,\n",
    "                    \"num_indiv\":num_individuals,\n",
    "                    \"A\":A,\n",
    "                    \"a\":a})\n",
    "    return ret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "trait_snpassoc_pca_county = pd.concat([trait_complete.countyid, trait_snpassoc_pca], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "trait_snpassoc_pca_county = trait_snpassoc_pca_county.drop(trait_snpassoc_pca_county[np.isnan(trait_snpassoc_pca_county[trait_name])].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "snpassoc_af = trait_snpassoc_pca_county.ix[:,2:-14].apply(get_allele_freqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0-10037-01-257</th>\n",
       "      <th>0-10040-02-394</th>\n",
       "      <th>0-10044-01-392</th>\n",
       "      <th>0-10048-01-60</th>\n",
       "      <th>0-10051-02-166</th>\n",
       "      <th>0-10054-01-402</th>\n",
       "      <th>0-10067-03-111</th>\n",
       "      <th>0-10079-02-168</th>\n",
       "      <th>0-10112-01-169</th>\n",
       "      <th>0-10113-01-119</th>\n",
       "      <th>0-10116-01-165</th>\n",
       "      <th>0-10151-01-86</th>\n",
       "      <th>0-10162-01-255</th>\n",
       "      <th>0-10207-01-280</th>\n",
       "      <th>0-10210-01-41</th>\n",
       "      <th>...</th>\n",
       "      <th>UMN-CL299Contig1-01-46</th>\n",
       "      <th>UMN-CL306Contig1-04-261</th>\n",
       "      <th>UMN-CL307Contig1-04-143</th>\n",
       "      <th>UMN-CL319Contig1-03-131</th>\n",
       "      <th>UMN-CL326Contig1-05-421</th>\n",
       "      <th>UMN-CL339Contig1-05-39</th>\n",
       "      <th>UMN-CL34Contig1-03-89</th>\n",
       "      <th>UMN-CL353Contig1-04-64</th>\n",
       "      <th>UMN-CL362Contig1-07-133</th>\n",
       "      <th>UMN-CL363Contig1-01-233</th>\n",
       "      <th>UMN-CL379Contig1-12-117</th>\n",
       "      <th>UMN-CL424Contig1-03-94</th>\n",
       "      <th>UMN-CL54Contig1-07-88</th>\n",
       "      <th>UMN-CL91Contig1-02-246</th>\n",
       "      <th>UMN-CL97Contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>          A</td>\n",
       "      <td>          C</td>\n",
       "      <td>          C</td>\n",
       "      <td>         G</td>\n",
       "      <td>          G</td>\n",
       "      <td>           A</td>\n",
       "      <td>          A</td>\n",
       "      <td>          G</td>\n",
       "      <td>          A</td>\n",
       "      <td>          A</td>\n",
       "      <td>          A</td>\n",
       "      <td>          A</td>\n",
       "      <td>           A</td>\n",
       "      <td>         A</td>\n",
       "      <td>          A</td>\n",
       "      <td>...</td>\n",
       "      <td>          A</td>\n",
       "      <td>           T</td>\n",
       "      <td>         C</td>\n",
       "      <td>          C</td>\n",
       "      <td>         G</td>\n",
       "      <td>           A</td>\n",
       "      <td>          G</td>\n",
       "      <td>           A</td>\n",
       "      <td>         C</td>\n",
       "      <td>          G</td>\n",
       "      <td>          A</td>\n",
       "      <td>          C</td>\n",
       "      <td>         G</td>\n",
       "      <td>         C</td>\n",
       "      <td>          G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fis</th>\n",
       "      <td> 0.03790158</td>\n",
       "      <td>-0.01800787</td>\n",
       "      <td> 0.03905122</td>\n",
       "      <td> 0.1123853</td>\n",
       "      <td> 0.07030303</td>\n",
       "      <td> 0.001200686</td>\n",
       "      <td>-0.03827751</td>\n",
       "      <td> 0.01592548</td>\n",
       "      <td>-0.06158833</td>\n",
       "      <td>-0.07300731</td>\n",
       "      <td>  0.1143453</td>\n",
       "      <td> 0.08119658</td>\n",
       "      <td> 0.001719146</td>\n",
       "      <td> 0.1537527</td>\n",
       "      <td>  0.1143967</td>\n",
       "      <td>...</td>\n",
       "      <td> 0.09555371</td>\n",
       "      <td> 0.007814833</td>\n",
       "      <td>-0.1143142</td>\n",
       "      <td> 0.05193962</td>\n",
       "      <td> 0.1273477</td>\n",
       "      <td> 0.009049774</td>\n",
       "      <td>-0.03865741</td>\n",
       "      <td>-0.006125574</td>\n",
       "      <td>-0.1328293</td>\n",
       "      <td> 0.07723564</td>\n",
       "      <td>-0.00921659</td>\n",
       "      <td> 0.03601537</td>\n",
       "      <td> 0.0345339</td>\n",
       "      <td>-0.1029759</td>\n",
       "      <td> 0.04555035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>He</th>\n",
       "      <td>  0.4138331</td>\n",
       "      <td>  0.4972566</td>\n",
       "      <td>  0.4999144</td>\n",
       "      <td> 0.4394845</td>\n",
       "      <td>  0.1290743</td>\n",
       "      <td>   0.5006011</td>\n",
       "      <td> 0.07385994</td>\n",
       "      <td> 0.09294359</td>\n",
       "      <td>  0.1120043</td>\n",
       "      <td>  0.4358706</td>\n",
       "      <td>  0.0448881</td>\n",
       "      <td>  0.1482612</td>\n",
       "      <td>   0.1047955</td>\n",
       "      <td> 0.4290435</td>\n",
       "      <td>  0.1140579</td>\n",
       "      <td>...</td>\n",
       "      <td> 0.05056321</td>\n",
       "      <td>   0.1940632</td>\n",
       "      <td> 0.2963679</td>\n",
       "      <td>  0.2628849</td>\n",
       "      <td>  0.302699</td>\n",
       "      <td>   0.0981527</td>\n",
       "      <td>  0.3779142</td>\n",
       "      <td>  0.01510504</td>\n",
       "      <td> 0.4190812</td>\n",
       "      <td>  0.1495905</td>\n",
       "      <td> 0.02108229</td>\n",
       "      <td>  0.3838553</td>\n",
       "      <td>  0.400328</td>\n",
       "      <td> 0.2733863</td>\n",
       "      <td>  0.3470169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ho</th>\n",
       "      <td>  0.3981481</td>\n",
       "      <td>  0.5062112</td>\n",
       "      <td>  0.4803922</td>\n",
       "      <td> 0.3900929</td>\n",
       "      <td>       0.12</td>\n",
       "      <td>         0.5</td>\n",
       "      <td> 0.07668712</td>\n",
       "      <td> 0.09146341</td>\n",
       "      <td>  0.1189024</td>\n",
       "      <td>  0.4676923</td>\n",
       "      <td> 0.03975535</td>\n",
       "      <td>  0.1362229</td>\n",
       "      <td>   0.1046154</td>\n",
       "      <td> 0.3630769</td>\n",
       "      <td>  0.1010101</td>\n",
       "      <td>...</td>\n",
       "      <td> 0.04573171</td>\n",
       "      <td>   0.1925466</td>\n",
       "      <td> 0.3302469</td>\n",
       "      <td>  0.2492308</td>\n",
       "      <td> 0.2641509</td>\n",
       "      <td>  0.09726444</td>\n",
       "      <td>  0.3925234</td>\n",
       "      <td>  0.01519757</td>\n",
       "      <td> 0.4747475</td>\n",
       "      <td>  0.1380368</td>\n",
       "      <td>  0.0212766</td>\n",
       "      <td>  0.3700306</td>\n",
       "      <td> 0.3865031</td>\n",
       "      <td> 0.3015385</td>\n",
       "      <td>  0.3312102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>        459</td>\n",
       "      <td>        349</td>\n",
       "      <td>        319</td>\n",
       "      <td>       436</td>\n",
       "      <td>        605</td>\n",
       "      <td>         330</td>\n",
       "      <td>        627</td>\n",
       "      <td>        624</td>\n",
       "      <td>        617</td>\n",
       "      <td>        442</td>\n",
       "      <td>        639</td>\n",
       "      <td>        594</td>\n",
       "      <td>         614</td>\n",
       "      <td>       448</td>\n",
       "      <td>        558</td>\n",
       "      <td>...</td>\n",
       "      <td>        639</td>\n",
       "      <td>         574</td>\n",
       "      <td>       531</td>\n",
       "      <td>        549</td>\n",
       "      <td>       518</td>\n",
       "      <td>         624</td>\n",
       "      <td>        480</td>\n",
       "      <td>         653</td>\n",
       "      <td>       417</td>\n",
       "      <td>        599</td>\n",
       "      <td>        651</td>\n",
       "      <td>        485</td>\n",
       "      <td>       472</td>\n",
       "      <td>       544</td>\n",
       "      <td>        488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PQ</th>\n",
       "      <td>        129</td>\n",
       "      <td>        163</td>\n",
       "      <td>        147</td>\n",
       "      <td>       126</td>\n",
       "      <td>         39</td>\n",
       "      <td>         162</td>\n",
       "      <td>         25</td>\n",
       "      <td>         30</td>\n",
       "      <td>         39</td>\n",
       "      <td>        152</td>\n",
       "      <td>         13</td>\n",
       "      <td>         44</td>\n",
       "      <td>          34</td>\n",
       "      <td>       118</td>\n",
       "      <td>         30</td>\n",
       "      <td>...</td>\n",
       "      <td>         15</td>\n",
       "      <td>          62</td>\n",
       "      <td>       107</td>\n",
       "      <td>         81</td>\n",
       "      <td>        84</td>\n",
       "      <td>          32</td>\n",
       "      <td>        126</td>\n",
       "      <td>           5</td>\n",
       "      <td>       141</td>\n",
       "      <td>         45</td>\n",
       "      <td>          7</td>\n",
       "      <td>        121</td>\n",
       "      <td>       126</td>\n",
       "      <td>        98</td>\n",
       "      <td>        104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>        189</td>\n",
       "      <td>        295</td>\n",
       "      <td>        293</td>\n",
       "      <td>       210</td>\n",
       "      <td>         45</td>\n",
       "      <td>         318</td>\n",
       "      <td>         25</td>\n",
       "      <td>         32</td>\n",
       "      <td>         39</td>\n",
       "      <td>        208</td>\n",
       "      <td>         15</td>\n",
       "      <td>         52</td>\n",
       "      <td>          36</td>\n",
       "      <td>       202</td>\n",
       "      <td>         36</td>\n",
       "      <td>...</td>\n",
       "      <td>         17</td>\n",
       "      <td>          70</td>\n",
       "      <td>       117</td>\n",
       "      <td>        101</td>\n",
       "      <td>       118</td>\n",
       "      <td>          34</td>\n",
       "      <td>        162</td>\n",
       "      <td>           5</td>\n",
       "      <td>       177</td>\n",
       "      <td>         53</td>\n",
       "      <td>          7</td>\n",
       "      <td>        169</td>\n",
       "      <td>       180</td>\n",
       "      <td>       106</td>\n",
       "      <td>        140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>          G</td>\n",
       "      <td>          A</td>\n",
       "      <td>          G</td>\n",
       "      <td>         A</td>\n",
       "      <td>          A</td>\n",
       "      <td>           G</td>\n",
       "      <td>          G</td>\n",
       "      <td>          A</td>\n",
       "      <td>          C</td>\n",
       "      <td>          G</td>\n",
       "      <td>          C</td>\n",
       "      <td>          T</td>\n",
       "      <td>           G</td>\n",
       "      <td>         C</td>\n",
       "      <td>          G</td>\n",
       "      <td>...</td>\n",
       "      <td>          C</td>\n",
       "      <td>           A</td>\n",
       "      <td>         G</td>\n",
       "      <td>          A</td>\n",
       "      <td>         A</td>\n",
       "      <td>           G</td>\n",
       "      <td>          C</td>\n",
       "      <td>           G</td>\n",
       "      <td>         A</td>\n",
       "      <td>          A</td>\n",
       "      <td>          G</td>\n",
       "      <td>          A</td>\n",
       "      <td>         A</td>\n",
       "      <td>         A</td>\n",
       "      <td>          A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>num_indiv</th>\n",
       "      <td>        324</td>\n",
       "      <td>        322</td>\n",
       "      <td>        306</td>\n",
       "      <td>       323</td>\n",
       "      <td>        325</td>\n",
       "      <td>         324</td>\n",
       "      <td>        326</td>\n",
       "      <td>        328</td>\n",
       "      <td>        328</td>\n",
       "      <td>        325</td>\n",
       "      <td>        327</td>\n",
       "      <td>        323</td>\n",
       "      <td>         325</td>\n",
       "      <td>       325</td>\n",
       "      <td>        297</td>\n",
       "      <td>...</td>\n",
       "      <td>        328</td>\n",
       "      <td>         322</td>\n",
       "      <td>       324</td>\n",
       "      <td>        325</td>\n",
       "      <td>       318</td>\n",
       "      <td>         329</td>\n",
       "      <td>        321</td>\n",
       "      <td>         329</td>\n",
       "      <td>       297</td>\n",
       "      <td>        326</td>\n",
       "      <td>        329</td>\n",
       "      <td>        327</td>\n",
       "      <td>       326</td>\n",
       "      <td>       325</td>\n",
       "      <td>        314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>  0.7083333</td>\n",
       "      <td>  0.5419255</td>\n",
       "      <td>  0.5212418</td>\n",
       "      <td> 0.6749226</td>\n",
       "      <td>  0.9307692</td>\n",
       "      <td>   0.5092593</td>\n",
       "      <td>  0.9616564</td>\n",
       "      <td>  0.9512195</td>\n",
       "      <td>  0.9405488</td>\n",
       "      <td>       0.68</td>\n",
       "      <td>  0.9770642</td>\n",
       "      <td>  0.9195046</td>\n",
       "      <td>   0.9446154</td>\n",
       "      <td> 0.6892308</td>\n",
       "      <td>  0.9393939</td>\n",
       "      <td>...</td>\n",
       "      <td>  0.9740854</td>\n",
       "      <td>   0.8913043</td>\n",
       "      <td> 0.8194444</td>\n",
       "      <td>  0.8446154</td>\n",
       "      <td> 0.8144654</td>\n",
       "      <td>   0.9483283</td>\n",
       "      <td>  0.7476636</td>\n",
       "      <td>   0.9924012</td>\n",
       "      <td> 0.7020202</td>\n",
       "      <td>  0.9187117</td>\n",
       "      <td>  0.9893617</td>\n",
       "      <td>  0.7415902</td>\n",
       "      <td> 0.7239264</td>\n",
       "      <td> 0.8369231</td>\n",
       "      <td>  0.7770701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>  0.2916667</td>\n",
       "      <td>  0.4580745</td>\n",
       "      <td>  0.4787582</td>\n",
       "      <td> 0.3250774</td>\n",
       "      <td> 0.06923077</td>\n",
       "      <td>   0.4907407</td>\n",
       "      <td> 0.03834356</td>\n",
       "      <td> 0.04878049</td>\n",
       "      <td> 0.05945122</td>\n",
       "      <td>       0.32</td>\n",
       "      <td> 0.02293578</td>\n",
       "      <td> 0.08049536</td>\n",
       "      <td>  0.05538462</td>\n",
       "      <td> 0.3107692</td>\n",
       "      <td> 0.06060606</td>\n",
       "      <td>...</td>\n",
       "      <td> 0.02591463</td>\n",
       "      <td>   0.1086957</td>\n",
       "      <td> 0.1805556</td>\n",
       "      <td>  0.1553846</td>\n",
       "      <td> 0.1855346</td>\n",
       "      <td>  0.05167173</td>\n",
       "      <td>  0.2523364</td>\n",
       "      <td> 0.007598784</td>\n",
       "      <td> 0.2979798</td>\n",
       "      <td> 0.08128834</td>\n",
       "      <td>  0.0106383</td>\n",
       "      <td>  0.2584098</td>\n",
       "      <td> 0.2760736</td>\n",
       "      <td> 0.1630769</td>\n",
       "      <td>  0.2229299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total_alleles</th>\n",
       "      <td>        648</td>\n",
       "      <td>        644</td>\n",
       "      <td>        612</td>\n",
       "      <td>       646</td>\n",
       "      <td>        650</td>\n",
       "      <td>         648</td>\n",
       "      <td>        652</td>\n",
       "      <td>        656</td>\n",
       "      <td>        656</td>\n",
       "      <td>        650</td>\n",
       "      <td>        654</td>\n",
       "      <td>        646</td>\n",
       "      <td>         650</td>\n",
       "      <td>       650</td>\n",
       "      <td>        594</td>\n",
       "      <td>...</td>\n",
       "      <td>        656</td>\n",
       "      <td>         644</td>\n",
       "      <td>       648</td>\n",
       "      <td>        650</td>\n",
       "      <td>       636</td>\n",
       "      <td>         658</td>\n",
       "      <td>        642</td>\n",
       "      <td>         658</td>\n",
       "      <td>       594</td>\n",
       "      <td>        652</td>\n",
       "      <td>        658</td>\n",
       "      <td>        654</td>\n",
       "      <td>       652</td>\n",
       "      <td>       650</td>\n",
       "      <td>        628</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows × 3082 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              0-10037-01-257 0-10040-02-394 0-10044-01-392 0-10048-01-60  \\\n",
       "A                          A              C              C             G   \n",
       "Fis               0.03790158    -0.01800787     0.03905122     0.1123853   \n",
       "He                 0.4138331      0.4972566      0.4999144     0.4394845   \n",
       "Ho                 0.3981481      0.5062112      0.4803922     0.3900929   \n",
       "P                        459            349            319           436   \n",
       "PQ                       129            163            147           126   \n",
       "Q                        189            295            293           210   \n",
       "a                          G              A              G             A   \n",
       "num_indiv                324            322            306           323   \n",
       "p                  0.7083333      0.5419255      0.5212418     0.6749226   \n",
       "q                  0.2916667      0.4580745      0.4787582     0.3250774   \n",
       "total_alleles            648            644            612           646   \n",
       "\n",
       "              0-10051-02-166 0-10054-01-402 0-10067-03-111 0-10079-02-168  \\\n",
       "A                          G              A              A              G   \n",
       "Fis               0.07030303    0.001200686    -0.03827751     0.01592548   \n",
       "He                 0.1290743      0.5006011     0.07385994     0.09294359   \n",
       "Ho                      0.12            0.5     0.07668712     0.09146341   \n",
       "P                        605            330            627            624   \n",
       "PQ                        39            162             25             30   \n",
       "Q                         45            318             25             32   \n",
       "a                          A              G              G              A   \n",
       "num_indiv                325            324            326            328   \n",
       "p                  0.9307692      0.5092593      0.9616564      0.9512195   \n",
       "q                 0.06923077      0.4907407     0.03834356     0.04878049   \n",
       "total_alleles            650            648            652            656   \n",
       "\n",
       "              0-10112-01-169 0-10113-01-119 0-10116-01-165 0-10151-01-86  \\\n",
       "A                          A              A              A             A   \n",
       "Fis              -0.06158833    -0.07300731      0.1143453    0.08119658   \n",
       "He                 0.1120043      0.4358706      0.0448881     0.1482612   \n",
       "Ho                 0.1189024      0.4676923     0.03975535     0.1362229   \n",
       "P                        617            442            639           594   \n",
       "PQ                        39            152             13            44   \n",
       "Q                         39            208             15            52   \n",
       "a                          C              G              C             T   \n",
       "num_indiv                328            325            327           323   \n",
       "p                  0.9405488           0.68      0.9770642     0.9195046   \n",
       "q                 0.05945122           0.32     0.02293578    0.08049536   \n",
       "total_alleles            656            650            654           646   \n",
       "\n",
       "              0-10162-01-255 0-10207-01-280 0-10210-01-41      ...        \\\n",
       "A                          A              A             A      ...         \n",
       "Fis              0.001719146      0.1537527     0.1143967      ...         \n",
       "He                 0.1047955      0.4290435     0.1140579      ...         \n",
       "Ho                 0.1046154      0.3630769     0.1010101      ...         \n",
       "P                        614            448           558      ...         \n",
       "PQ                        34            118            30      ...         \n",
       "Q                         36            202            36      ...         \n",
       "a                          G              C             G      ...         \n",
       "num_indiv                325            325           297      ...         \n",
       "p                  0.9446154      0.6892308     0.9393939      ...         \n",
       "q                 0.05538462      0.3107692    0.06060606      ...         \n",
       "total_alleles            650            650           594      ...         \n",
       "\n",
       "              UMN-CL299Contig1-01-46 UMN-CL306Contig1-04-261  \\\n",
       "A                                  A                       T   \n",
       "Fis                       0.09555371             0.007814833   \n",
       "He                        0.05056321               0.1940632   \n",
       "Ho                        0.04573171               0.1925466   \n",
       "P                                639                     574   \n",
       "PQ                                15                      62   \n",
       "Q                                 17                      70   \n",
       "a                                  C                       A   \n",
       "num_indiv                        328                     322   \n",
       "p                          0.9740854               0.8913043   \n",
       "q                         0.02591463               0.1086957   \n",
       "total_alleles                    656                     644   \n",
       "\n",
       "              UMN-CL307Contig1-04-143 UMN-CL319Contig1-03-131  \\\n",
       "A                                   C                       C   \n",
       "Fis                        -0.1143142              0.05193962   \n",
       "He                          0.2963679               0.2628849   \n",
       "Ho                          0.3302469               0.2492308   \n",
       "P                                 531                     549   \n",
       "PQ                                107                      81   \n",
       "Q                                 117                     101   \n",
       "a                                   G                       A   \n",
       "num_indiv                         324                     325   \n",
       "p                           0.8194444               0.8446154   \n",
       "q                           0.1805556               0.1553846   \n",
       "total_alleles                     648                     650   \n",
       "\n",
       "              UMN-CL326Contig1-05-421 UMN-CL339Contig1-05-39  \\\n",
       "A                                   G                      A   \n",
       "Fis                         0.1273477            0.009049774   \n",
       "He                           0.302699              0.0981527   \n",
       "Ho                          0.2641509             0.09726444   \n",
       "P                                 518                    624   \n",
       "PQ                                 84                     32   \n",
       "Q                                 118                     34   \n",
       "a                                   A                      G   \n",
       "num_indiv                         318                    329   \n",
       "p                           0.8144654              0.9483283   \n",
       "q                           0.1855346             0.05167173   \n",
       "total_alleles                     636                    658   \n",
       "\n",
       "              UMN-CL34Contig1-03-89 UMN-CL353Contig1-04-64  \\\n",
       "A                                 G                      A   \n",
       "Fis                     -0.03865741           -0.006125574   \n",
       "He                        0.3779142             0.01510504   \n",
       "Ho                        0.3925234             0.01519757   \n",
       "P                               480                    653   \n",
       "PQ                              126                      5   \n",
       "Q                               162                      5   \n",
       "a                                 C                      G   \n",
       "num_indiv                       321                    329   \n",
       "p                         0.7476636              0.9924012   \n",
       "q                         0.2523364            0.007598784   \n",
       "total_alleles                   642                    658   \n",
       "\n",
       "              UMN-CL362Contig1-07-133 UMN-CL363Contig1-01-233  \\\n",
       "A                                   C                       G   \n",
       "Fis                        -0.1328293              0.07723564   \n",
       "He                          0.4190812               0.1495905   \n",
       "Ho                          0.4747475               0.1380368   \n",
       "P                                 417                     599   \n",
       "PQ                                141                      45   \n",
       "Q                                 177                      53   \n",
       "a                                   A                       A   \n",
       "num_indiv                         297                     326   \n",
       "p                           0.7020202               0.9187117   \n",
       "q                           0.2979798              0.08128834   \n",
       "total_alleles                     594                     652   \n",
       "\n",
       "              UMN-CL379Contig1-12-117 UMN-CL424Contig1-03-94  \\\n",
       "A                                   A                      C   \n",
       "Fis                       -0.00921659             0.03601537   \n",
       "He                         0.02108229              0.3838553   \n",
       "Ho                          0.0212766              0.3700306   \n",
       "P                                 651                    485   \n",
       "PQ                                  7                    121   \n",
       "Q                                   7                    169   \n",
       "a                                   G                      A   \n",
       "num_indiv                         329                    327   \n",
       "p                           0.9893617              0.7415902   \n",
       "q                           0.0106383              0.2584098   \n",
       "total_alleles                     658                    654   \n",
       "\n",
       "              UMN-CL54Contig1-07-88 UMN-CL91Contig1-02-246 UMN-CL97Contig  \n",
       "A                                 G                      C              G  \n",
       "Fis                       0.0345339             -0.1029759     0.04555035  \n",
       "He                         0.400328              0.2733863      0.3470169  \n",
       "Ho                        0.3865031              0.3015385      0.3312102  \n",
       "P                               472                    544            488  \n",
       "PQ                              126                     98            104  \n",
       "Q                               180                    106            140  \n",
       "a                                 A                      A              A  \n",
       "num_indiv                       326                    325            314  \n",
       "p                         0.7239264              0.8369231      0.7770701  \n",
       "q                         0.2760736              0.1630769      0.2229299  \n",
       "total_alleles                   652                    650            628  \n",
       "\n",
       "[12 rows x 3082 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpassoc_af"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "getting allele freqs for pop  1\n",
      "getting allele freqs for pop  2\n",
      "getting allele freqs for pop  3\n",
      "getting allele freqs for pop  5\n",
      "getting allele freqs for pop  6\n",
      "getting allele freqs for pop  7\n",
      "getting allele freqs for pop  8\n",
      "getting allele freqs for pop  9\n",
      "getting allele freqs for pop  11\n",
      "getting allele freqs for pop  12\n",
      "getting allele freqs for pop  13\n",
      "getting allele freqs for pop  14\n",
      "getting allele freqs for pop  16\n",
      "getting allele freqs for pop  17\n",
      "getting allele freqs for pop  18\n",
      "getting allele freqs for pop  19\n",
      "getting allele freqs for pop  20\n",
      "getting allele freqs for pop  21\n",
      "getting allele freqs for pop  23\n",
      "getting allele freqs for pop  24\n",
      "getting allele freqs for pop  25\n",
      "getting allele freqs for pop  26\n",
      "getting allele freqs for pop  27\n",
      "getting allele freqs for pop  28\n",
      "getting allele freqs for pop  29\n",
      "getting allele freqs for pop  30\n",
      "getting allele freqs for pop  31\n",
      "getting allele freqs for pop  32\n",
      "getting allele freqs for pop  33\n",
      "getting allele freqs for pop  34\n"
     ]
    }
   ],
   "source": [
    "pop_allele_freqs = {}\n",
    "for pop,data in trait_snpassoc_pca_county.groupby(\"countyid\"):\n",
    "    print \"getting allele freqs for pop % d\" % pop\n",
    "    pop_allele_freqs[pop] = data.ix[:,2:-14].apply(get_allele_freqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_usable_counties(county):\n",
    "    if county.county_state in county_id:\n",
    "        return True\n",
    "    return False\n",
    "\n",
    "data_ai = hdf.get(\"data_ai\")\n",
    "data_ai['usable'] = data_ai.apply(get_usable_counties, axis=1)\n",
    "data_ai = data_ai.drop(data_ai[data_ai.usable == False].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def write_gwas_data_file(df, pheno, outdir):\n",
    "    out = \"%s_gwas_data_file.txt\" % pheno\n",
    "    out = os.path.join(outdir, out)\n",
    "    df = df.sort_index()\n",
    "    df[['A1', 'A2', 'EFF', 'FRQ']].to_csv(out,\n",
    "                                          header=True, \n",
    "                                          index=True,\n",
    "                                          sep=\"\\t\")\n",
    "    print out\n",
    "    return out\n",
    "\n",
    "def write_freqs_file(df, pheno, pop_freqs, outdir):\n",
    "    out = \"%s_freqs_file.txt\" % pheno\n",
    "    out = os.path.join(outdir, out)\n",
    "    print out\n",
    "    with open(out, \"w\") as o:\n",
    "        o.write(\"SNP\\tCLST\\tA1\\tA2\\tFRQ\\n\")\n",
    "        for pop, data in pop_freqs.items():\n",
    "            m = data.T.merge(df, how=\"inner\", left_index=True, right_index=True)\n",
    "            m['population'] = pop\n",
    "            m.index.name = 'SNP'\n",
    "            m = m.sort_index()\n",
    "            o.write(m[['population','A1','A2','p']].to_csv(header=False, \n",
    "                                                             index=True,\n",
    "                                                             sep=\"\\t\"))\n",
    "def write_match_pop_file(df, pheno, pop_freqs, pop, outdir):\n",
    "    out = \"%s_match_pop_file.txt\" % pheno\n",
    "    out = os.path.join(outdir, out)\n",
    "    print out\n",
    "    with open(out, \"w\") as o:\n",
    "        o.write(\"SNP\\tCLST\\tA1\\tA2\\tFRQ\\n\")\n",
    "        for key, data in pop_freqs.items():\n",
    "            if key == pop:\n",
    "                m = data.T.merge(df, how=\"inner\", left_index=True, right_index=True)\n",
    "                m['population'] = pop\n",
    "                m.index.name = 'SNP'\n",
    "                m = m.sort_index()\n",
    "                o.write(m[['population','A1','A2','p']].to_csv(header=False, \n",
    "                                                                 index=True,\n",
    "                                                                 sep=\"\\t\"))\n",
    "                break\n",
    "                \n",
    "def write_full_dataset_file(df, pheno, pop_freqs, outdir):\n",
    "    out = \"%s_full_dataset_file.txt\" % pheno\n",
    "    out = os.path.join(outdir, out)\n",
    "    print out\n",
    "    with open(out, \"w\") as o:\n",
    "        o.write(\"SNP\\tCLST\\tA1\\tA2\\tFRQ\\n\")\n",
    "        for pop, data in pop_freqs.items():\n",
    "            m = data.T.merge(df, how=\"inner\", left_index=True, right_index=True)\n",
    "            m['population'] = pop\n",
    "            m.index.name = 'SNP'\n",
    "            m = m.sort_index()\n",
    "            o.write(m[['population','A1','A2','p']].to_csv(header=False, \n",
    "                                                             index=True,\n",
    "                                                             sep=\"\\t\"))   \n",
    "def write_env_var_data_file(pheno, pop_freqs, outdir):\n",
    "    cols = [x for x in data_ai.columns if \"AI\" in x]\n",
    "    for c in cols:\n",
    "        pop_id = 1\n",
    "        out = \"%s_%s_env_var_data_file.txt\" % (c, pheno)\n",
    "        out = os.path.join(outdir, out)\n",
    "        print out\n",
    "        with open(out, \"w\") as o:\n",
    "            o.write(\"CLST\\tENV\\tREG\\n\")\n",
    "            for pop, d in data_ai.groupby('county_state'):\n",
    "                pop = float(county_id[pop])*1.0\n",
    "                if pop in pop_freqs:\n",
    "                    o.write(\"%.1f\\t%f\\t%.1f\\n\" % (pop, d[c], pop_id))\n",
    "                    pop_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/gdc_home4/cfried/ipython\n"
     ]
    }
   ],
   "source": [
    "cd \"~/ipython/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chose 2 candidates\n",
      "squat_cfried/melezitose_gwas_data_file.txt\n",
      "squat_cfried/melezitose_freqs_file.txt\n",
      "squat_cfried/melezitose_match_pop_file.txt\n",
      "squat_cfried/melezitose_full_dataset_file.txt\n",
      "squat_cfried/AI_Q1_melezitose_env_var_data_file.txt\n",
      "squat_cfried/AI_Q2_melezitose_env_var_data_file.txt\n",
      "squat_cfried/AI_Q3_melezitose_env_var_data_file.txt\n",
      "squat_cfried/AI_Q4_melezitose_env_var_data_file.txt\n"
     ]
    }
   ],
   "source": [
    "squat_outdir = \"squat_cfried\" #change for your username\n",
    "if not os.path.exists(squat_outdir):\n",
    "    os.mkdir(squat_outdir)\n",
    "\n",
    "for p in alpha_vals:\n",
    "    full = alpha_vals[p].T\n",
    "    full['q-value'] = multipletests(full['p-value'], method=\"fdr_bh\")[1]\n",
    "    full.index = [x.replace(\".\", \"-\") for x in full.index]\n",
    "    full.index = [x[1:] if x.startswith(\"X\") else x for x in full.index]\n",
    "    full.index.name = \"SNP\"\n",
    "    full.AA = full.AA.apply(lambda x: x[0])\n",
    "    full.aa = full.aa.apply(lambda x: x[0])\n",
    "    full = full.rename(columns={'alpha':'EFF',\n",
    "                                'AA':'A1',\n",
    "                                'aa':'A2',\n",
    "                                'p': 'FRQ'})\n",
    "    candidates = full[full['q-value']<0.05]\n",
    "    print \"chose %d candidates\" % len(candidates)\n",
    "    write_gwas_data_file(candidates, p, squat_outdir)\n",
    "    write_freqs_file(candidates, p, pop_allele_freqs, squat_outdir)\n",
    "    write_match_pop_file(full, p, pop_allele_freqs, 2, squat_outdir)\n",
    "    write_full_dataset_file(full, p, pop_allele_freqs, squat_outdir)\n",
    "    write_env_var_data_file(p, pop_allele_freqs, squat_outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['squat_cfried/AI_Q1_melezitose_env_var_data_file.txt', 'squat_cfried/AI_Q2_melezitose_env_var_data_file.txt', 'squat_cfried/AI_Q3_melezitose_env_var_data_file.txt', 'squat_cfried/AI_Q4_melezitose_env_var_data_file.txt']\n"
     ]
    }
   ],
   "source": [
    "env_squat_files = !find {squat_outdir} | grep {trait_name} | grep env_var | grep AI\n",
    "print env_squat_files\n",
    "env_squat_files = [os.path.basename(x) for x in env_squat_files]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "env_var_file_string = \"list(%s)\" % \", \".join([\"'%s'\" % x for x in env_squat_files])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "squat_cfried/squat_melezitose.r\n",
      "system(\"rm -rf melezitose\")\n",
      "source('/gdc_home4/cfried/src/PolygenicAdaptationCode/Scripts/CreateTraitFile.R')\n",
      "source('/gdc_home4/cfried/src/PolygenicAdaptationCode/Scripts/functions.R')\n",
      "PolygenicAdaptationFunction(sim.null=T,\n",
      "cov.SNPs.per.cycle=5000,\n",
      "null.cycles=1,\n",
      "match.bins=list(seq(0,0.5,0.02), c(2), seq(0,1000,100)),\n",
      "load.cov.mat=F,\n",
      "path='melezitose',\n",
      "cov.cycles=1,\n",
      "match.pop.file='melezitose_match_pop_file.txt',\n",
      "freqs.file='melezitose_freqs_file.txt',\n",
      "env.var.data.files=list('AI_Q1_melezitose_env_var_data_file.txt', 'AI_Q2_melezitose_env_var_data_file.txt', 'AI_Q3_melezitose_env_var_data_file.txt', 'AI_Q4_melezitose_env_var_data_file.txt'),\n",
      "gwas.data.file='melezitose_gwas_data_file.txt',\n",
      "full.dataset.file='melezitose_full_dataset_file.txt',\n",
      "match.categories=c('MAF'),\n",
      "check.allele.orientation=F,\n",
      "null.phenos.per.cycle=1000)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "squat_scripts_dir = \"/gdc_home4/cfried/src/PolygenicAdaptationCode/Scripts\"\n",
    "!rm {squat_outdir}/Scripts && ln -s {squat_scripts_dir} {squat_outdir}/Scripts\n",
    "def get_squat_vars(pheno):\n",
    "    d = {\"gwas.data.file\":\"'%s_gwas_data_file.txt'\" % pheno,\n",
    "         \"freqs.file\":\"'%s_freqs_file.txt'\" % pheno,\n",
    "         \"env.var.data.files\": env_var_file_string,\n",
    "         \"match.pop.file\":\"'%s_match_pop_file.txt'\" % pheno,\n",
    "         \"full.dataset.file\":\"'%s_full_dataset_file.txt'\" % pheno,\n",
    "         \"path\":\"'%s'\" % pheno,\n",
    "         \"match.categories\":\"c('MAF')\",\n",
    "         \"match.bins\":\"list(seq(0,0.5,0.02), c(2), seq(0,1000,100))\",\n",
    "         \"cov.SNPs.per.cycle\":5000,\n",
    "         \"cov.cycles\":1,\n",
    "         \"null.phenos.per.cycle\":1000,\n",
    "         \"null.cycles\":1,\n",
    "         \"load.cov.mat\":\"F\",\n",
    "         \"sim.null\":\"T\",\n",
    "         \"check.allele.orientation\":\"F\"}\n",
    "    return ',\\n'.join(\"%s=%s\" % (key,val) for (key,val) in d.items())\n",
    "\n",
    "def create_squat_run_file(pheno):\n",
    "    squat_file = os.path.join(squat_outdir, \"squat_%s.r\" % pheno)\n",
    "    with open(squat_file, \"w\") as o:\n",
    "        o.write('system(\"rm -rf %s\")\\n'% pheno)\n",
    "        o.write(\"source('%s')\\n\" % os.path.join(squat_scripts_dir, \"CreateTraitFile.R\"))\n",
    "        o.write(\"source('%s')\\n\" % os.path.join(squat_scripts_dir, \"functions.R\"))\n",
    "        o.write(\"PolygenicAdaptationFunction(%s)\\n\" % get_squat_vars(pheno))\n",
    "    return squat_file\n",
    "\n",
    "for pheno in alpha_vals:\n",
    "    squat_file = create_squat_run_file(pheno)\n",
    "    print squat_file\n",
    "    !cat $squat_file\n",
    "    print \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running melezitose\n",
      "setwd('squat_cfried')\n",
      "source(\"squat_melezitose.r\")\n",
      "setwd('../')\n"
     ]
    }
   ],
   "source": [
    "def run_squat(p):\n",
    "    print \"running %s\" % p\n",
    "    output = \"%s/%s\" % (squat_outdir, p)\n",
    "    if os.path.exists(output):\n",
    "        !rm -rf {output}\n",
    "    cmds = [\"setwd('%s')\" % squat_outdir,\n",
    "            'source(\"squat_%s.r\")' % (p),\n",
    "            \"setwd('../')\"]\n",
    "    for cmd in cmds:\n",
    "        print cmd\n",
    "        r(cmd)\n",
    "    \n",
    "run_squat(trait_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'melezitose': ['squat_cfried/melezitose/Output/genetic.values.Robj',\n",
       "  'squat_cfried/melezitose/Output/theStats.Robj',\n",
       "  'squat_cfried/melezitose/Output/asymptotic.pVals.Robj',\n",
       "  'squat_cfried/melezitose/Output/pVals.Robj',\n",
       "  'squat_cfried/melezitose/Output/nullStats.Robj']}"
      ]
     },
     "execution_count": 325,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rfiles = !find {squat_outdir} | grep Robj | grep Output | grep {trait_name}\n",
    "bc = {}\n",
    "for f in rfiles:\n",
    "    d = f.split(\"/\")\n",
    "    if not d[1] in bc:\n",
    "        bc[d[1]] = []\n",
    "    bc[d[1]].append(f)\n",
    "bc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "melezitose\n",
      "$Qx\n",
      "[1] 23.99343\n",
      "\n",
      "$Fst.comp\n",
      "[1] 31.37303\n",
      "\n",
      "$LD.component\n",
      "[1] -7.379601\n",
      "\n",
      "$betas\n",
      "$betas[[1]]\n",
      "[1] -0.001558947\n",
      "\n",
      "$betas[[2]]\n",
      "[1] -0.02474625\n",
      "\n",
      "$betas[[3]]\n",
      "[1] -0.002661235\n",
      "\n",
      "$betas[[4]]\n",
      "[1] -0.004253784\n",
      "\n",
      "\n",
      "$pearson.rs\n",
      "$pearson.rs[[1]]\n",
      "[1] -0.2800495\n",
      "\n",
      "$pearson.rs[[2]]\n",
      "[1] -0.2884058\n",
      "\n",
      "$pearson.rs[[3]]\n",
      "[1] -0.06345363\n",
      "\n",
      "$pearson.rs[[4]]\n",
      "[1] -0.2173006\n",
      "\n",
      "\n",
      "$spearman.rhos\n",
      "$spearman.rhos[[1]]\n",
      "           [,1]\n",
      "[1,] -0.2098522\n",
      "\n",
      "$spearman.rhos[[2]]\n",
      "           [,1]\n",
      "[1,] -0.3108374\n",
      "\n",
      "$spearman.rhos[[3]]\n",
      "            [,1]\n",
      "[1,] -0.08768473\n",
      "\n",
      "$spearman.rhos[[4]]\n",
      "           [,1]\n",
      "[1,] -0.2285714\n",
      "\n",
      "\n",
      "$reg.Z\n",
      "           Env File 1  Env File 2  Env File 3  Env File 4\n",
      "Region 1  -0.81623269 -0.81623269 -0.81623269 -0.81623269\n",
      "Region 2  -2.23817714 -2.23817714 -2.23817714 -2.23817714\n",
      "Region 3   0.92708918  0.92708918  0.92708918  0.92708918\n",
      "Region 4   0.48331684  0.48331684  0.48331684  0.48331684\n",
      "Region 5   0.10726352  0.10726352  0.10726352  0.10726352\n",
      "Region 6   0.36322870  0.36322870  0.36322870  0.36322870\n",
      "Region 7  -1.57282923 -1.57282923 -1.57282923 -1.57282923\n",
      "Region 8   0.83740719  0.83740719  0.83740719  0.83740719\n",
      "Region 9   0.37069364  0.37069364  0.37069364  0.37069364\n",
      "Region 10  1.14483327  1.14483327  1.14483327  1.14483327\n",
      "Region 11  0.03690586  0.03690586  0.03690586  0.03690586\n",
      "Region 12 -0.62420107 -0.62420107 -0.62420107 -0.62420107\n",
      "Region 13  0.43382072  0.43382072  0.43382072  0.43382072\n",
      "Region 14  0.33094585  0.33094585  0.33094585  0.33094585\n",
      "Region 15 -0.96294505 -0.96294505 -0.96294505 -0.96294505\n",
      "Region 16 -0.17642106 -0.17642106 -0.17642106 -0.17642106\n",
      "Region 17 -0.54734458 -0.54734458 -0.54734458 -0.54734458\n",
      "Region 18 -0.21425110 -0.21425110 -0.21425110 -0.21425110\n",
      "Region 19  0.64752713  0.64752713  0.64752713  0.64752713\n",
      "Region 20  0.54319545  0.54319545  0.54319545  0.54319545\n",
      "Region 21  0.70751600  0.70751600  0.70751600  0.70751600\n",
      "Region 22  0.12753502  0.12753502  0.12753502  0.12753502\n",
      "Region 23  0.74744595  0.74744595  0.74744595  0.74744595\n",
      "Region 24  0.10952570  0.10952570  0.10952570  0.10952570\n",
      "Region 25 -1.36584110 -1.36584110 -1.36584110 -1.36584110\n",
      "Region 26  1.09072860  1.09072860  1.09072860  1.09072860\n",
      "Region 27  1.43052581  1.43052581  1.43052581  1.43052581\n",
      "Region 28 -0.22034194 -0.22034194 -0.22034194 -0.22034194\n",
      "Region 29  1.59602371  1.59602371  1.59602371  1.59602371\n",
      "Region 30 -0.79187468 -0.79187468 -0.79187468 -0.79187468\n",
      "\n",
      "$ind.Z\n",
      "          1           2           3           5           6           7 \n",
      "-0.81623269 -2.23817714  0.92708918  0.48331684  0.10726352  0.36322870 \n",
      "          8           9          11          12          13          14 \n",
      "-1.57282923  0.83740719  0.37069364  1.14483327  0.03690586 -0.62420107 \n",
      "         16          17          18          19          20          21 \n",
      " 0.43382072  0.33094585 -0.96294505 -0.17642106 -0.54734458 -0.21425110 \n",
      "         23          24          25          26          27          28 \n",
      " 0.64752713  0.54319545  0.70751600  0.12753502  0.74744595  0.10952570 \n",
      "         29          30          31          32          33          34 \n",
      "-1.36584110  1.09072860  1.43052581 -0.22034194  1.59602371 -0.79187468 \n",
      "\n",
      "\n",
      "------------------\n",
      "$Qx\n",
      "[1] 0.36\n",
      "\n",
      "$Fst.comp\n",
      "[1] 0.046\n",
      "\n",
      "$LD.comp\n",
      "[1] 0.946\n",
      "\n",
      "$betas\n",
      "[1] 0.694 0.716 0.992 0.722\n",
      "\n",
      "$pearson.rs\n",
      "[1] 0.181 0.099 0.780 0.321\n",
      "\n",
      "$spearman.rhos\n",
      "[1] 0.370 0.142 0.762 0.284\n",
      "\n",
      "$reg.Z\n",
      "          Env File 1 Env File 2 Env File 3 Env File 4\n",
      "Region 1       0.260      0.260      0.260      0.260\n",
      "Region 2       0.004      0.004      0.004      0.004\n",
      "Region 3       0.278      0.278      0.278      0.278\n",
      "Region 4       0.758      0.758      0.758      0.758\n",
      "Region 5       0.874      0.874      0.874      0.874\n",
      "Region 6       0.752      0.752      0.752      0.752\n",
      "Region 7       0.032      0.032      0.032      0.032\n",
      "Region 8       0.386      0.386      0.386      0.386\n",
      "Region 9       0.818      0.818      0.818      0.818\n",
      "Region 10      0.282      0.282      0.282      0.282\n",
      "Region 11      0.892      0.892      0.892      0.892\n",
      "Region 12      0.630      0.630      0.630      0.630\n",
      "Region 13      0.486      0.486      0.486      0.486\n",
      "Region 14      0.918      0.918      0.918      0.918\n",
      "Region 15      0.436      0.436      0.436      0.436\n",
      "Region 16      0.698      0.698      0.698      0.698\n",
      "Region 17      0.426      0.426      0.426      0.426\n",
      "Region 18      0.982      0.982      0.982      0.982\n",
      "Region 19      0.468      0.468      0.468      0.468\n",
      "Region 20      0.628      0.628      0.628      0.628\n",
      "Region 21      0.512      0.512      0.512      0.512\n",
      "Region 22      0.836      0.836      0.836      0.836\n",
      "Region 23      0.314      0.314      0.314      0.314\n",
      "Region 24      0.880      0.880      0.880      0.880\n",
      "Region 25      0.108      0.108      0.108      0.108\n",
      "Region 26      0.164      0.164      0.164      0.164\n",
      "Region 27      0.150      0.150      0.150      0.150\n",
      "Region 28      0.964      0.964      0.964      0.964\n",
      "Region 29      0.064      0.064      0.064      0.064\n",
      "Region 30      0.328      0.328      0.328      0.328\n",
      "\n",
      "$ind.Z\n",
      "    1     2     3     5     6     7     8     9    11    12    13    14    16 \n",
      "0.260 0.004 0.278 0.758 0.874 0.752 0.032 0.386 0.818 0.282 0.892 0.630 0.486 \n",
      "   17    18    19    20    21    23    24    25    26    27    28    29    30 \n",
      "0.918 0.436 0.698 0.426 0.982 0.468 0.628 0.512 0.836 0.314 0.880 0.108 0.164 \n",
      "   31    32    33    34 \n",
      "0.150 0.964 0.064 0.328 \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for pheno in bc:\n",
    "    print pheno\n",
    "    for obj in bc[pheno]:\n",
    "        r('load(\"%s\")' % obj)\n",
    "    print r(\"the.stats\")\n",
    "    print(\"------------------\")\n",
    "    print r(\"p.vals\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
