%!TEX root = slides.tex

\section[Session 1]{Session 1: Evolutionary genetics of associations among
genotypes, phenotypes, and environments}

\begin{frame}
\frametitle{Before we get started, let's set the stage for where we are}
\begin{block}{}
\centering
\includegraphics[height=0.8\textheight]{sork}\\
\tiny
Adapted from \citet[Figure 1]{Sork:2013tb}
\end{block}{}
\end{frame}

\begin{frame}
\frametitle{What I'm not going to talk about (mostly)}
\begin{block}{Let's avoid}
\begin{itemize}
\item{A thorough review of quantitative genetics}
\item{A thorough review of population genetics}
\item{Unnecessary derivations of any formulae}
\item{Computer programming, systems engineering, setup, etc.}
\end{itemize}
\end{block}
\end{frame}

\begin{frame}
\frametitle{What I'm going to talk about (mostly)}
\begin{block}{Instead, let's try}
\begin{itemize}
\item{Fixed and random effects, linear models, linear mixed models}
\item{Enough quant/pop gen to get by}
\item{Some basic equations}
\item{Genetic variance partitioning}
\item{Some experimental design decisions}
\item{Data types, methods and tools}
\item{Experiments in both natural and transplant populations}
\end{itemize}
\end{block}
\end{frame}

\begin{frame}
\frametitle{Fixed vs random effects}
\begin{block}{Fixed effects}
\begin{itemize}
\item{\textbf{Treatment} effects}
\item{Case vs. control, and \textbf{only} these two groups}
\item{Variable for which all levels are included (e.g., gender, race)}
\item{Explicitly compare between levels without generalizing (e.g., San
Francisco vs. Berkeley)}
\item{Usually qualitiative variables}
\end{itemize}
\end{block}
\tiny
\citet{StroupFreund200203}
\end{frame}

\begin{frame}
\frametitle{Fixed vs random effects}
\begin{block}{Random effects}
\begin{itemize}
\item{Drawn from a distribution of possible effects}
\item{San Francisco and Berkeley are just two of many western US, 
or liberal, or hippie, or measels-infected cities}
\item{Use of a random effect allows for more generalizable conclusions 
from the data}
\item{rather than comparing among levels for an effect, random 
effects measure how much variance in the dependent variable is 
accounted for across levels of the random factor}
\item{Blocking, control, repeated-measures factors are random}
\end{itemize}
\end{block}
\tiny
\citet{StroupFreund200203}
\end{frame}

\begin{frame}
\frametitle{Evoluationary quantitative genetics \citep{Walsh:2008}}
\begin{block}{Goals}
\begin{itemize}
\item{Understand the genetics and inheritance of complex traits}
\item{Nature and strength of evolutionary forces}
\item{Natural populations}
\end{itemize}
\end{block}


\begin{block}{Natural selection}
\begin{itemize}
\item{On traits}
\item{Heritability}
\item{Variation/Variance}
\item{Additive genetic variation}
\end{itemize}
\end{block}
\end{frame}

\begin{frame}
\frametitle{The data we have is from multiple sources}
\begin{block}{Our data}
\begin{itemize}
\item{A sampling of individuals}
\item{A sampling of populations made up of those individuals}
\item{Some genetic data from these individuals}
\item{Some phenotypic data about the individuals}
\item{Some data about the populations (e.g., location)}
\end{itemize}

\end{block}\end{frame}

\begin{frame}
\frametitle{There are several things we'd like to do with our data}

\begin{block}{Genetic architecture}
\begin{itemize}
\item{Ascertain a meaningful set of genetic variants which sufficient
discriminatory power}
\item{Use this genetic variation to understand something about the 
traits of the organism we care about}
\item{In other words, given a set of traits that we observe in nature, and as
scientists we find interesting, can we attribute variation in that trait with
variation in the genome}
\end{itemize}
\end{block}
\end{frame}

\begin{frame}
\frametitle{There is some math we need}
\begin{block}{}
\begin{equation}
\label{eqn:V_P}
V_P = V_G + V_E + V_{GE}
\end{equation}

\begin{equation}
\label{eqn:V_G}
V_G = V_A + V_D + V_I
\end{equation}

\begin{equation}
\label{eqn:V_A}
V_A = 2pq\alpha^2
\end{equation}

\begin{equation}
\label{eqn:alpha}
\alpha = a + d(q-p)
\end{equation}

\begin{equation}
\label{eqn:a}
a = \frac{G_{AA}-G_{aa}}{2}
\end{equation}

\begin{equation}
\label{eqn:d}
d = G_{Aa} - \frac{G_{AA}+G_{aa}}{2}
\end{equation}

\end{block}{}
\end{frame}

\begin{frame}
\frametitle{Variance components}
\begin{block}{Equation \ref{eqn:V_P}: $V_P = V_G + V_E + V_{GE}$}
\begin{itemize}
\item{$V_P$: Phenotypic variance ($\sigma^2_P$)}
\item{$V_G$: Genetic variance}
\item{$V_E$: Environmental variance}
\item{$V_{GE}$: Genetic*environment}
\end{itemize}
\end{block}
\end{frame}

\begin{frame}
\frametitle{Genetic variance}
\begin{block}{Equation \ref{eqn:V_G}: $V_G = V_A + V_D + V_I$}
\begin{itemize}
\item{$V_G$: Total genetic variance}
\item{$V_A$: Additive genetic variance}
\item{$V_D$: Dominance genetic variance}
\item{$V_I$: Epistatic genetic variance (I = interaction)}
\end{itemize}
\end{block}
\end{frame}

\begin{frame}
\frametitle{Additive gentic variance}
\begin{block}{}
\begin{itemize}
\item{Contribution of these alleles to a phenotype are independent of 1) other
genes and 2) the environment}
\item{When multiple alleles contribute to a single phenotype (polygenic), their
presence has a linear effect on the phenotype}
\item{$V_A$ is the target of natural selection}
\end{itemize}
\end{block}{}
\end{frame}

\begin{frame}
\frametitle{Heritability}
\begin{block}{The Breeder's equation}
\begin{center}
\huge
$R = h^2S$
\end{center}
\begin{itemize}
\item[]{$R$: response to selection}
\item[]{$h^2$: narrow sense heritability ($\frac{V_A}{V_P}$)}
\item[]{$S$: selection coefficient}
\end{itemize}
\end{block}

\begin{block}{Example}
\href{http://localhost:8888/notebooks/heritability.ipynb}{heratibility.ipynb}
\end{block}
\end{frame}



\begin{frame}
\frametitle{Local adaptation}
\begin{block}{}
\centering
\includegraphics[width=0.75\textwidth]{salvo_fig1}\\
\tiny
Adapted from \citet[Figure 1]{Savolainen:2013dfa}
\end{block}
\end{frame}

\begin{frame}
\frametitle{Methods to detect local adaptation}
\begin{block}{Current approaches*}{}
\begin{itemize}
\item{QTL-mapping}
\item{Population genetics}
\item{Association mapping}
\end{itemize}
\end{block}
\tiny
*Adapted from \citet{Savolainen:2013dfa}
\end{frame}

\begin{frame}
\frametitle{Methods to detect local adaptation}
\begin{block}{QTL}
\begin{itemize}
\item{Can operate on a reduced set of markers (GBS: RAD, ddRAD)}
\item{Does not require a reference genome}
\item{Need a linkage map, large sample sizes, many crosses}
\end{itemize}
\end{block}

\begin{block}{Popgen}
\begin{itemize}
\item{Diverse data sets (SNPs, microsatellites): DNA sequence variation,
$F_{ST}$, $Q_{ST}$}
\item{Conculsions driven by data type}
\item{Requires a reference genome}
\item{Normal challenges of next-gen data apply (quality filtering, redundancy,
coverage)}
\end{itemize}
\end{block}
\end{frame}

\begin{frame}
\frametitle{Methods to detect local adaptation}
\begin{block}{Association Mapping}
\begin{itemize}
\item{Sampling from populations}
\item{Much lower LD among loci}
\item{Needs dense sets of markers (varies with LD and genome size)}
\item{Must consider genetically differentiated populations, population
structure is important!}
\end{itemize}
\end{block}
\end{frame}

\begin{frame}
\frametitle{Association mapping in conifers}
\begin{block}{GWAS in trees}
\begin{itemize}
\item{Forest trees are important economically and environmentally}
\item{Studying them in traditional ways (i.e., QTL) is difficult: generation
time is long, phenotype not present in seedlings (e.g, wood/bark)}
\item{Complex demography, population structure, local adaptation can confound 
associations}
\item{Heritability tends to be low for traits of interest}
\item{Mapping populations may not generalize to natural populations}
\end{itemize}
\end{block}
\tiny
\citet{Uchiyama:2013ci}
\end{frame}

\begin{frame}
\frametitle{Association mapping in conifers}
\begin{block}{Conifer genomes}
\begin{itemize}
\item{Huge: \SIrange{10}{30}{GB}}
\item{Highy complex (repeats, gene content, GC-bias)}
\item{Ratio of genetic to physical distance $>$\SI{3000}{kb \per cM}}
\item{Rapid decay of LD in coding regions}
\end{itemize}
\end{block}
\tiny
\citet{Uchiyama:2013ci,Hirschhorn:2005cka}
\end{frame}

\begin{frame}
\frametitle{Association mapping in conifers}
\begin{block}{Identifying variants}
Two main methods:
\begin{itemize}
\item{Genotyping arrays}
\item{Gentoyping by sequencing (GBS)}
\item{Each have own set of biases and problems}
\end{itemize}
\end{block}
\end{frame}

\begin{frame}
\frametitle{Now that you have identified variants...}
\begin{block}{Let's revisit what you want to do}
\centering
You want to understand how the variation in your biologically important and
interesting trait is attributable to the underlying additive genetic variation 
present (i.e., heritability) in your samples (and hopefully the population).
\end{block}
\end{frame}

\begin{frame}
\frametitle{Now that you have identified variants...}
\begin{block}{}
\centering
\includegraphics[width=0.9\textwidth]{hirschhorn_table1}\\
\tiny
\citet[Table 1]{Hirschhorn:2005cka}
\end{block}
\end{frame}

\begin{frame}
\frametitle{Testing SNPs against phenotypes}
\begin{block}{}
\begin{center}
\Large{$y=X \bm{\beta} + S \bm{\alpha} + Q \bm{v} + Z \bm{u} + \bm{e}$}
\end{center}
\begin{itemize}
\item{$X \beta$: fixed effects w/o SNP + population structure}
\item{$\beta$: fixed effects other than SNP or pop. structure}
\item{$\alpha$: vector of SNP effects}
\item{$v$: vector of population effects}
\item{$u$: vector of polygene background effects}
\item{$Q$: matrix from STRUCTURE relating y to v}
\item{$X, S, Z$: incidence matrices (0/1) relating y to $\beta$, $\alpha$, $u$,
respectively}
\end{itemize}
\end{block}
\tiny
\citet{Yu:2006ij}
\end{frame}

\begin{frame}
\frametitle{Testing SNPs against phenotypes}
\begin{block}{}
\begin{center}
\Large{$y=X \bm{\beta} + S \bm{\alpha} + Q \bm{v} + Z \bm{u} + \bm{e}$}
\end{center}
\begin{itemize}
\item{$Var(u) = 2KV_g, Var(e) = RV_R$}
\item{$K$: $n \times n$ matrix of genetic covariance between pairs of
individuals (kinship)}
\item{$R$: $n \times n$ matrix off-diag elems are 0, diag is
$1/N_{{obs}_{pheno}}$}
\item{$V_g$: genetic variance (additive)}
\item{$V_R$: residual variance (non-additive + env)}
\end{itemize}
\end{block}
\tiny
\citet{Yu:2006ij}
\end{frame}